From c47b9e04253828b8da00108185e9151b7662a1ab Mon Sep 17 00:00:00 2001
From: Philippe Gerum <rpm@xenomai.org>
Date: Thu, 23 Apr 2020 15:52:51 +0200
Subject: [PATCH 0649/1010] evl: introduce the observable element

Observables enable the observer design pattern, in which any number of
observer threads can be notified of updates to any number of
observable subjects, in a loosely coupled fashion. In the same move,
an EVL thread becomes in and of itself an observable which can be
monitored for events.

As a by-product, provide support for channeling SIGDEBUG notifications
to threads through their own observable on request instead of, or in
addition to issuing SIGDEBUG. The built-in runtime error detection
capabilities the core provides can feed health monitors, based on the
observability of threads.

The core switches to ABI 23 as a result of these changes.

Signed-off-by: Philippe Gerum <rpm@xenomai.org>
---
 include/asm-generic/evl/thread_info.h |    3 +
 include/evl/factory.h                 |   51 +-
 include/evl/mutex.h                   |    4 +-
 include/evl/observable.h              |   59 ++
 include/evl/poll.h                    |    1 +
 include/evl/thread.h                  |   46 +-
 include/evl/wait.h                    |   21 +-
 include/trace/events/evl.h            |   45 +-
 include/uapi/evl/control.h            |    4 +-
 include/uapi/evl/factory.h            |   12 +-
 include/uapi/evl/observable.h         |   45 +
 include/uapi/evl/poll.h               |    2 +
 include/uapi/evl/signal.h             |   13 +-
 include/uapi/evl/thread.h             |   28 +-
 include/uapi/evl/types.h              |   10 +
 kernel/evl/Kconfig                    |   27 +-
 kernel/evl/Makefile                   |    1 +
 kernel/evl/factory.c                  |   47 +-
 kernel/evl/monitor.c                  |   19 +-
 kernel/evl/mutex.c                    |   44 +-
 kernel/evl/observable.c               | 1147 +++++++++++++++++++++++++
 kernel/evl/poll.c                     |    5 +
 kernel/evl/proxy.c                    |    7 +-
 kernel/evl/sched/core.c               |   29 +-
 kernel/evl/stax.c                     |    3 +-
 kernel/evl/syscall.c                  |    8 +-
 kernel/evl/thread.c                   |  293 +++++--
 kernel/evl/wait.c                     |   10 +-
 kernel/evl/xbuf.c                     |    3 +
 29 files changed, 1749 insertions(+), 238 deletions(-)
 create mode 100644 include/evl/observable.h
 create mode 100644 include/uapi/evl/observable.h
 create mode 100644 kernel/evl/observable.c

diff --git a/include/asm-generic/evl/thread_info.h b/include/asm-generic/evl/thread_info.h
index ed2bd105f5..2aa2a697a9 100644
--- a/include/asm-generic/evl/thread_info.h
+++ b/include/asm-generic/evl/thread_info.h
@@ -5,9 +5,11 @@
 #ifdef CONFIG_EVL
 
 struct evl_thread;
+struct evl_subscriber;
 
 struct oob_thread_state {
 	struct evl_thread *thread;
+	struct evl_subscriber *subscriber;
 	int preempt_count;
 };
 
@@ -15,6 +17,7 @@ static inline
 void evl_init_thread_state(struct oob_thread_state *p)
 {
 	p->thread = NULL;
+	p->subscriber = NULL;
 	p->preempt_count = 0;
 }
 
diff --git a/include/evl/factory.h b/include/evl/factory.h
index 02ee9e52a0..7238acf000 100644
--- a/include/evl/factory.h
+++ b/include/evl/factory.h
@@ -38,6 +38,12 @@ struct evl_element;
 
 #define EVL_DEVHASH_BITS	8
 
+struct evl_index {
+	struct rb_root root;
+	hard_spinlock_t lock;
+	fundle_t generator;
+};
+
 struct evl_factory {
 	const char *name;
 	const struct file_operations *fops;
@@ -59,11 +65,7 @@ struct evl_factory {
 		kuid_t kuid;
 		kgid_t kgid;
 		unsigned long *minor_map;
-		struct evl_index {
-			struct rb_root root;
-			hard_spinlock_t lock;
-			fundle_t generator;
-		} index;
+		struct evl_index index;
 		DECLARE_HASHTABLE(name_hash, EVL_DEVHASH_BITS);
 		struct mutex hash_lock;
 	}; /* Internal. */
@@ -109,16 +111,22 @@ void evl_destroy_element(struct evl_element *e);
 void evl_get_element(struct evl_element *e);
 
 struct evl_element *
-__evl_get_element_by_fundle(struct evl_factory *fac,
+__evl_get_element_by_fundle(struct evl_index *map,
 			fundle_t fundle);
 
-#define evl_get_element_by_fundle(__fac, __fundle, __type)		\
+#define evl_get_element_by_fundle(__map, __fundle, __type)		\
 	({								\
 		struct evl_element *__e;				\
-		__e = __evl_get_element_by_fundle(__fac, __fundle);	\
+		__e = __evl_get_element_by_fundle(__map, __fundle);	\
 		__e ? container_of(__e, __type, element) : NULL;	\
 	})
 
+#define evl_get_factory_element_by_fundle(__fac, __fundle, __type)	\
+	({								\
+		struct evl_index *__map = &(__fac)->index;		\
+		evl_get_element_by_fundle(__map, __fundle, __type);	\
+	})
+
 /*
  * An element can be disposed of only after the device backing it is
  * removed. If @dev is valid, so is @e at the time of the call.
@@ -135,9 +143,14 @@ static inline bool evl_element_is_public(struct evl_element *e)
 	return !!(e->clone_flags & EVL_CLONE_PUBLIC);
 }
 
-static inline bool evl_element_is_core(struct evl_element *e)
+static inline bool evl_element_has_coredev(struct evl_element *e)
+{
+	return !!(e->clone_flags & EVL_CLONE_COREDEV);
+}
+
+static inline bool evl_element_is_observable(struct evl_element *e)
 {
-	return !!(e->clone_flags & EVL_CLONE_CORE);
+	return !!(e->clone_flags & EVL_CLONE_OBSERVABLE);
 }
 
 void evl_put_element(struct evl_element *e);
@@ -154,12 +167,21 @@ int evl_create_core_element_device(struct evl_element *e,
 
 void evl_remove_element_device(struct evl_element *e);
 
-void evl_index_element(struct evl_element *e);
+void evl_index_element(struct evl_index *map,
+		struct evl_element *e);
 
-int evl_index_element_at(struct evl_element *e,
-			fundle_t fundle);
+static inline void evl_index_factory_element(struct evl_element *e)
+{
+	evl_index_element(&e->factory->index, e);
+}
 
-void evl_unindex_element(struct evl_element *e);
+void evl_unindex_element(struct evl_index *map,
+			struct evl_element *e);
+
+static inline void evl_unindex_factory_element(struct evl_element *e)
+{
+	evl_unindex_element(&e->factory->index, e);
+}
 
 int evl_early_init_factories(void);
 
@@ -177,5 +199,6 @@ extern struct evl_factory evl_thread_factory;
 extern struct evl_factory evl_trace_factory;
 extern struct evl_factory evl_xbuf_factory;
 extern struct evl_factory evl_proxy_factory;
+extern struct evl_factory evl_observable_factory;
 
 #endif /* !_EVL_FACTORY_H */
diff --git a/include/evl/mutex.h b/include/evl/mutex.h
index 14a81bd6b2..35ee0f1c7f 100644
--- a/include/evl/mutex.h
+++ b/include/evl/mutex.h
@@ -13,7 +13,7 @@
 #include <evl/list.h>
 #include <evl/assert.h>
 #include <evl/timer.h>
-#include <evl/thread.h>
+#include <evl/wait.h>
 #include <uapi/evl/mutex.h>
 
 struct evl_clock;
@@ -79,7 +79,7 @@ void evl_flush_mutex(struct evl_mutex *mutex,
 
 void evl_commit_mutex_ceiling(struct evl_mutex *mutex);
 
-void evl_detect_boost_drop(struct evl_thread *owner);
+void evl_detect_boost_drop(void);
 
 int evl_reorder_mutex_wait(struct evl_thread *waiter,
 			struct evl_thread *originator);
diff --git a/include/evl/observable.h b/include/evl/observable.h
new file mode 100644
index 0000000000..f0766a4455
--- /dev/null
+++ b/include/evl/observable.h
@@ -0,0 +1,59 @@
+/*
+ * SPDX-License-Identifier: GPL-2.0
+ *
+ * Copyright (C) 2020 Philippe Gerum  <rpm@xenomai.org>
+ */
+
+#ifndef _EVL_OBSERVABLE_H
+#define _EVL_OBSERVABLE_H
+
+#include <linux/types.h>
+#include <linux/ktime.h>
+#include <linux/list.h>
+#include <linux/wait.h>
+#include <evl/factory.h>
+#include <evl/wait.h>
+#include <evl/poll.h>
+#include <uapi/evl/observable.h>
+
+struct file;
+
+struct evl_observable {
+	struct evl_element element;
+	struct list_head observers; 	/* struct evl_observer */
+	struct list_head flush_list; 	/* struct evl_observer */
+	struct evl_wait_queue oob_wait;
+	wait_queue_head_t inband_wait;
+	struct evl_poll_head poll_head;
+	struct irq_work wake_irqwork;
+	struct irq_work flush_irqwork;
+	evl_spinlock_t lock;		/* guards observers and flush_list */
+	u32 serial_counter;
+	int writable_observers;
+};
+
+void evl_drop_subscriptions(struct evl_subscriber *subscriber);
+
+struct evl_observable *evl_alloc_observable(int clone_flags);
+
+void evl_flush_observable(struct evl_observable *observable);
+
+bool evl_send_observable(struct evl_observable *observable, int tag,
+			union evl_value details);
+
+ssize_t evl_read_observable(struct evl_observable *observable,
+			char __user *u_buf, size_t count, bool wait);
+
+ssize_t evl_write_observable(struct evl_observable *observable,
+			const char __user *u_buf, size_t count);
+
+__poll_t evl_oob_poll_observable(struct evl_observable *observable,
+				struct oob_poll_wait *wait);
+
+__poll_t evl_poll_observable(struct evl_observable *observable,
+			struct file *filp, poll_table *pt);
+
+long evl_ioctl_observable(struct evl_observable *observable,
+			unsigned int cmd, unsigned long arg);
+
+#endif /* !_EVL_OBSERVABLE_H */
diff --git a/include/evl/poll.h b/include/evl/poll.h
index 522229a355..e9125a98be 100644
--- a/include/evl/poll.h
+++ b/include/evl/poll.h
@@ -41,6 +41,7 @@ struct evl_poll_watchpoint {
 	unsigned int fd;
 	int events_polled;
 	int events_received;
+	union evl_value pollval;
 	struct oob_poll_wait wait;
 	struct evl_flag *flag;
 	struct file *filp;
diff --git a/include/evl/thread.h b/include/evl/thread.h
index 291f774881..05f41a474e 100644
--- a/include/evl/thread.h
+++ b/include/evl/thread.h
@@ -31,35 +31,32 @@
 #define EVL_THREAD_BLOCK_BITS   (T_SUSP|T_PEND|T_DELAY|T_WAIT|T_DORMANT|T_INBAND|T_HALT|T_PTSYNC)
 /* Information bits an EVL thread may receive from a blocking op. */
 #define EVL_THREAD_INFO_MASK	(T_RMID|T_TIMEO|T_BREAK|T_WAKEN|T_ROBBED|T_KICKED|T_BCAST)
+/* Mode bits configurable via EVL_THRIOC_SET/CLEAR_MODE. */
+#define EVL_THREAD_MODE_BITS	(T_WOSS|T_WOLI|T_WOSX|T_HMSIG|T_HMOBS)
 
 /*
- * These are special internal values of SIGDEBUG causes which are
- * never sent to user-space, but specifically handled by
- * evl_switch_inband().
+ * These are special internal values of HM diags which are never sent
+ * to user-space, but specifically handled by evl_switch_inband().
  */
-#define SIGDEBUG_NONE   0
-#define SIGDEBUG_TRAP  -1
+#define EVL_HMDIAG_NONE   0
+#define EVL_HMDIAG_TRAP  -1
 
 struct evl_thread;
 struct evl_rq;
 struct evl_sched_class;
 struct evl_poll_watchpoint;
+struct evl_wait_channel;
+struct evl_observable;
+struct file;
 
 struct evl_init_thread_attr {
 	const struct cpumask *affinity;
+	struct evl_observable *observable;
 	int flags;
 	struct evl_sched_class *sched_class;
 	union evl_sched_param sched_param;
 };
 
-struct evl_wait_channel {
-	int (*reorder_wait)(struct evl_thread *waiter,
-			struct evl_thread *originator);
-	int (*follow_depend)(struct evl_wait_channel *wchan,
-			struct evl_thread *originator);
-	struct list_head wait_list;
-};
-
 struct evl_thread {
 	evl_spinlock_t lock;
 
@@ -154,8 +151,9 @@ struct evl_thread {
 	struct completion exited;
 	kernel_cap_t raised_cap;
 	struct list_head kill_next;
-	struct oob_mm_state *oob_mm; /* Mostly RO. */
-	struct list_head ptsync_next; /* covered by oob_mm->lock. */
+	struct oob_mm_state *oob_mm;	/* Mostly RO. */
+	struct list_head ptsync_next;	/* covered by oob_mm->lock. */
+	struct evl_observable *observable;
 	char *name;
 };
 
@@ -237,6 +235,16 @@ static inline void evl_test_cancel(void)
 		__evl_test_cancel(curr);
 }
 
+static inline struct evl_subscriber *evl_get_subscriber(void)
+{
+	return dovetail_current_state()->subscriber;
+}
+
+static inline void evl_set_subscriber(struct evl_subscriber *sbr)
+{
+	dovetail_current_state()->subscriber = sbr;
+}
+
 ktime_t evl_get_thread_timeout(struct evl_thread *thread);
 
 ktime_t evl_get_thread_period(struct evl_thread *thread);
@@ -283,7 +291,10 @@ int evl_wait_thread_period(unsigned long *overruns_r);
 void evl_cancel_thread(struct evl_thread *thread);
 
 int evl_join_thread(struct evl_thread *thread,
-		bool uninterruptible);
+		    bool uninterruptible);
+
+void evl_notify_thread(struct evl_thread *thread,
+		       int tag, union evl_value details);
 
 void evl_get_thread_state(struct evl_thread *thread,
 			struct evl_thread_state *statebuf);
@@ -308,6 +319,8 @@ int evl_set_thread_schedparam(struct evl_thread *thread,
 
 int evl_killall(int mask);
 
+bool evl_is_thread_file(struct file *filp);
+
 void __evl_propagate_schedparam_change(struct evl_thread *curr);
 
 static inline void evl_propagate_schedparam_change(struct evl_thread *curr)
@@ -325,6 +338,7 @@ int __evl_run_kthread(struct evl_kthread *kthread, int clone_flags);
 		struct evl_init_thread_attr __iattr = {			\
 			.flags = 0,					\
 			.affinity = __affinity,				\
+			.observable = NULL,				\
 			.sched_class = &evl_sched_fifo,			\
 			.sched_param.fifo.prio = __priority,		\
 		};							\
diff --git a/include/evl/wait.h b/include/evl/wait.h
index 59891d7771..d8f27f7ed1 100644
--- a/include/evl/wait.h
+++ b/include/evl/wait.h
@@ -11,17 +11,24 @@
 #include <linux/types.h>
 #include <linux/errno.h>
 #include <linux/spinlock.h>
+#include <evl/lock.h>
 #include <evl/list.h>
 #include <evl/timer.h>
 #include <evl/clock.h>
-#include <evl/thread.h>
-#include <evl/sched.h>
 #include <trace/events/evl.h>
 #include <uapi/evl/thread.h>
 
 #define EVL_WAIT_FIFO    0
 #define EVL_WAIT_PRIO    BIT(0)
 
+struct evl_wait_channel {
+	int (*reorder_wait)(struct evl_thread *waiter,
+			struct evl_thread *originator);
+	int (*follow_depend)(struct evl_wait_channel *wchan,
+			struct evl_thread *originator);
+	struct list_head wait_list;
+};
+
 struct evl_wait_queue {
 	int flags;
 	struct evl_clock *clock;
@@ -90,14 +97,6 @@ static inline bool evl_wait_active(struct evl_wait_queue *wq)
 	return !list_empty(&wq->wchan.wait_list);
 }
 
-static inline
-struct evl_thread *evl_wait_head(struct evl_wait_queue *wq)
-{
-	assert_evl_lock(&wq->lock);
-	return list_first_entry_or_null(&wq->wchan.wait_list,
-					struct evl_thread, wait_next);
-}
-
 void __evl_init_wait(struct evl_wait_queue *wq,
 		struct evl_clock *clock,
 		int flags,
@@ -112,6 +111,8 @@ void __evl_init_wait(struct evl_wait_queue *wq,
 
 void evl_destroy_wait(struct evl_wait_queue *wq);
 
+struct evl_thread *evl_wait_head(struct evl_wait_queue *wq);
+
 void evl_flush_wait_locked(struct evl_wait_queue *wq,
 			int reason);
 
diff --git a/include/trace/events/evl.h b/include/trace/events/evl.h
index 4c62b98358..c1cde0faa5 100644
--- a/include/trace/events/evl.h
+++ b/include/trace/events/evl.h
@@ -293,7 +293,7 @@ TRACE_EVENT(evl_init_thread,
 	TP_fast_assign(
 		__entry->thread = thread;
 		__assign_str(thread_name, thread->name);
-		__entry->flags = iattr->flags;
+		__entry->flags = iattr->flags | (iattr->observable ? T_OBSERV : 0);
 		__assign_str(class_name, iattr->sched_class->name);
 		__entry->cprio = thread->cprio;
 		__entry->status = status;
@@ -497,18 +497,18 @@ DEFINE_EVENT(curr_thread_event, evl_switched_oob,
 	TP_ARGS(curr)
 );
 
-#define evl_print_switch_cause(cause)					\
-	__print_symbolic(cause,						\
-			{ SIGDEBUG_TRAP,		"breakpoint trap" }, \
-			{ SIGDEBUG_NONE,		"undefined" },	\
-			{ SIGDEBUG_MIGRATE_SIGNAL,	"signal" },	\
-			{ SIGDEBUG_MIGRATE_SYSCALL,	"syscall" },	\
-			{ SIGDEBUG_MIGRATE_FAULT,	"fault" },	\
-			{ SIGDEBUG_MIGRATE_PRIOINV,     "priority inversion" }, \
-			{ SIGDEBUG_WATCHDOG,		"watchdog" },	\
-			{ SIGDEBUG_MUTEX_IMBALANCE,     "mutex imbalance" }, \
-			{ SIGDEBUG_MUTEX_SLEEP,         "mutex sleep" }, \
-			{ SIGDEBUG_STAGE_LOCKED,        "stage exclusion" } )
+#define evl_print_switch_cause(cause)						\
+	__print_symbolic(cause,							\
+			{ EVL_HMDIAG_TRAP,		"breakpoint trap" },	\
+			{ EVL_HMDIAG_NONE,		"undefined" },		\
+			{ EVL_HMDIAG_SIGDEMOTE,		"in-band signal" }, 	\
+			{ EVL_HMDIAG_SYSDEMOTE,		"in-band syscall" },	\
+			{ EVL_HMDIAG_EXDEMOTE,		"processor exception" },\
+			{ EVL_HMDIAG_WATCHDOG,		"watchdog" },		\
+			{ EVL_HMDIAG_LKDEPEND,		"lock dependency" },	\
+			{ EVL_HMDIAG_LKIMBALANCE,	"lock imbalance" },	\
+			{ EVL_HMDIAG_LKSLEEP,		"sleep holding lock" },	\
+			{ EVL_HMDIAG_STAGEX,		"stage exclusion" } )
 
 TRACE_EVENT(evl_switch_inband,
 	TP_PROTO(int cause),
@@ -579,23 +579,26 @@ TRACE_EVENT(evl_inband_wakeup,
 );
 
 TRACE_EVENT(evl_inband_signal,
-	TP_PROTO(struct task_struct *task, int sig),
-	TP_ARGS(task, sig),
+	TP_PROTO(struct evl_thread *thread, int sig, int sigval),
+	TP_ARGS(thread, sig, sigval),
 
 	TP_STRUCT__entry(
-		__field(pid_t, pid)
-		__array(char, comm, TASK_COMM_LEN)
+		__field(struct evl_thread *, thread)
 		__field(int, sig)
+		__field(int, sigval)
 	),
 
 	TP_fast_assign(
-		__entry->pid = task_pid_nr(task);
+		__entry->thread = thread;
 		__entry->sig = sig;
-		memcpy(__entry->comm, task->comm, TASK_COMM_LEN);
+		__entry->sigval = sigval;
 	),
 
-	TP_printk("pid=%d comm=%s sig=%d",
-		  __entry->pid, __entry->comm, __entry->sig)
+	/* Caller holds a reference on @thread, memory cannot be stale. */
+	TP_printk("thread=%s pid=%d sig=%d sigval=%d",
+		evl_element_name(&__entry->thread->element),
+		evl_get_inband_pid(__entry->thread),
+		__entry->sig, __entry->sigval)
 );
 
 DEFINE_EVENT(timer_event, evl_timer_stop,
diff --git a/include/uapi/evl/control.h b/include/uapi/evl/control.h
index 5117423d06..794f8137de 100644
--- a/include/uapi/evl/control.h
+++ b/include/uapi/evl/control.h
@@ -11,14 +11,14 @@
 #include <uapi/evl/sched.h>
 
 /* Earliest ABI level we support. */
-#define EVL_ABI_BASE   21
+#define EVL_ABI_BASE   23
 /*
  * Current/latest ABI level we support. We may decouple the base and
  * current ABI levels by providing backward compatibility from the
  * latter to the former. CAUTION: a litteral value is required for the
  * current ABI definition (scripts reading this may be naive).
  */
-#define EVL_ABI_LEVEL  22
+#define EVL_ABI_LEVEL  23
 
 #define EVL_CONTROL_DEV  "/dev/evl/control"
 
diff --git a/include/uapi/evl/factory.h b/include/uapi/evl/factory.h
index 6e5235499b..3f255bcda7 100644
--- a/include/uapi/evl/factory.h
+++ b/include/uapi/evl/factory.h
@@ -17,10 +17,14 @@ struct evl_element_ids {
 	__u32 state_offset;
 };
 
-#define EVL_CLONE_PUBLIC  (1 << 16)
-#define EVL_CLONE_PRIVATE (0 << 16)
-#define EVL_CLONE_CORE    (1 << 31)
-#define EVL_CLONE_MASK    ((__u32)-1 << 16)
+/* The core only uses bits 16-31, rest is available to libevl. */
+#define EVL_CLONE_PUBLIC	(1 << 16)
+#define EVL_CLONE_PRIVATE	(0 << 16)
+#define EVL_CLONE_OBSERVABLE	(1 << 17)
+#define EVL_CLONE_NONBLOCK	(1 << 18)
+#define EVL_CLONE_MASTER	(1 << 19)
+#define EVL_CLONE_COREDEV	(1 << 31)
+#define EVL_CLONE_MASK		(((__u32)-1 << 16) & ~EVL_CLONE_COREDEV)
 
 struct evl_clone_req {
 	__u64 name_ptr;		/* (const char *name) */
diff --git a/include/uapi/evl/observable.h b/include/uapi/evl/observable.h
new file mode 100644
index 0000000000..5e610bde4b
--- /dev/null
+++ b/include/uapi/evl/observable.h
@@ -0,0 +1,45 @@
+/*
+ * SPDX-License-Identifier: GPL-2.0 WITH Linux-syscall-note
+ *
+ * Copyright (C) 2020 Philippe Gerum  <rpm@xenomai.org>
+ */
+
+#ifndef _EVL_UAPI_OBSERVABLE_H
+#define _EVL_UAPI_OBSERVABLE_H
+
+#include <uapi/evl/types.h>
+
+#define EVL_OBSERVABLE_DEV	"observable"
+
+/* __evl_notification.flags */
+#define EVL_NOTIFY_ALWAYS	(0 << 0)
+#define EVL_NOTIFY_ONCHANGE	(1 << 0)
+#define EVL_NOTIFY_MASK		EVL_NOTIFY_ONCHANGE
+
+struct evl_notice {
+	__u32 tag;
+	union evl_value event;
+};
+
+/* Notice tags below this value are reserved to the core. */
+#define EVL_NOTICE_USER  64
+
+struct evl_subscription {
+	__u32 backlog_count;
+	__u32 flags;
+};
+
+struct __evl_notification {
+	__u32 tag;
+	__u32 serial;
+	__s32 issuer;
+	union evl_value event;
+	struct __evl_timespec date;
+};
+
+#define EVL_OBSERVABLE_IOCBASE	'o'
+
+#define EVL_OBSIOC_SUBSCRIBE		_IOW(EVL_OBSERVABLE_IOCBASE, 0, struct evl_subscription)
+#define EVL_OBSIOC_UNSUBSCRIBE		_IO(EVL_OBSERVABLE_IOCBASE, 1)
+
+#endif /* !_EVL_UAPI_OBSERVABLE_H */
diff --git a/include/uapi/evl/poll.h b/include/uapi/evl/poll.h
index bb14c1edb7..321ebd016d 100644
--- a/include/uapi/evl/poll.h
+++ b/include/uapi/evl/poll.h
@@ -21,11 +21,13 @@ struct evl_poll_ctlreq {
 	__u32 action;
 	__u32 fd;
 	__u32 events;
+	union evl_value pollval;
 };
 
 struct evl_poll_event {
 	__u32 fd;
 	__u32 events;
+	union evl_value pollval;
 };
 
 struct evl_poll_waitreq {
diff --git a/include/uapi/evl/signal.h b/include/uapi/evl/signal.h
index f8f4f6b4db..3d4a869e1a 100644
--- a/include/uapi/evl/signal.h
+++ b/include/uapi/evl/signal.h
@@ -9,6 +9,9 @@
 #ifndef _EVL_UAPI_SIGNAL_H
 #define _EVL_UAPI_SIGNAL_H
 
+/*
+ * EVL_HMDIAG_xxx codes are possible values of sigdebug_cause().
+ */
 #define SIGDEBUG			SIGXCPU
 #define sigdebug_code(si)		((si)->si_value.sival_int)
 #define sigdebug_cause(si)		(sigdebug_code(si) & 0xff)
@@ -16,14 +19,4 @@
 #define sigdebug_marked(si)		\
 	((sigdebug_code(si) & 0xffff0000) == sigdebug_marker)
 
-/* Possible values of sigdebug_cause() */
-#define SIGDEBUG_MIGRATE_SIGNAL		1
-#define SIGDEBUG_MIGRATE_SYSCALL	2
-#define SIGDEBUG_MIGRATE_FAULT		3
-#define SIGDEBUG_MIGRATE_PRIOINV	4
-#define SIGDEBUG_WATCHDOG		5
-#define SIGDEBUG_MUTEX_IMBALANCE	6
-#define SIGDEBUG_MUTEX_SLEEP		7
-#define SIGDEBUG_STAGE_LOCKED		8
-
 #endif /* !_EVL_UAPI_SIGNAL_H */
diff --git a/include/uapi/evl/thread.h b/include/uapi/evl/thread.h
index fe5c3dd907..6152904652 100644
--- a/include/uapi/evl/thread.h
+++ b/include/uapi/evl/thread.h
@@ -30,10 +30,13 @@
 #define T_ROOT    0x00001000 /* Root thread (in-band kernel placeholder) */
 #define T_WEAK    0x00002000 /* Weak scheduling (in-band) */
 #define T_USER    0x00004000 /* Userland thread */
-#define T_WOSS    0x00008000 /* Warn on stage switch (SIGDEBUG) */
-#define T_WOLI    0x00010000 /* Warn on locking inconsistency (SIGDEBUG) */
-#define T_WOSX    0x00020000 /* Warn on stage exclusion (SIGDEBUG) */
+#define T_WOSS    0x00008000 /* Warn on stage switch (HM) */
+#define T_WOLI    0x00010000 /* Warn on locking inconsistency (HM) */
+#define T_WOSX    0x00020000 /* Warn on stage exclusion (HM) */
 #define T_PTRACE  0x00040000 /* Stopped on ptrace event */
+#define T_OBSERV  0x00080000 /* Observable (only for export to userland) */
+#define T_HMSIG   0x00100000 /* Notify HM events via SIGDEBUG */
+#define T_HMOBS   0x00200000 /* Notify HM events via observable */
 
 /* Information flags (shared) */
 
@@ -44,11 +47,11 @@
 #define T_WAKEN   0x00000010 /* Thread waken up upon resource availability */
 #define T_ROBBED  0x00000020 /* Robbed from resource ownership */
 #define T_CANCELD 0x00000040 /* Cancellation request is pending */
-#define T_PIALERT 0x00000080 /* Priority inversion alert (SIGDEBUG sent) */
+#define T_PIALERT 0x00000080 /* Priority inversion alert (HM notified) */
 #define T_SCHEDP  0x00000100 /* Schedparam propagation is pending */
 #define T_BCAST   0x00000200 /* Woken up upon resource broadcast */
 #define T_SIGNAL  0x00000400 /* Event monitor signaled */
-#define T_SXALERT 0x00000800 /* Stage exclusion alert (SIGDEBUG sent) */
+#define T_SXALERT 0x00000800 /* Stage exclusion alert (HM notified) */
 #define T_PTSIG   0x00001000 /* Ptrace signal is pending */
 #define T_PTSTOP  0x00002000 /* Ptrace stop is ongoing */
 #define T_PTJOIN  0x00004000 /* Ptracee should join ptsync barrier */
@@ -75,10 +78,21 @@
  * 'b' -> Priority boost undergoing
  * '#' -> Ptrace sync ongoing
  * 'r' -> Undergoes round-robin
- * 't' -> Warned on stage switch (T_WOSS -> SIGDEBUG)
+ * 't' -> Warned on stage switch (T_WOSS)
  * 'T' -> Stopped on ptrace event
+ * 'o' -> Observable
  */
-#define EVL_THREAD_STATE_LABELS  "SWDpRUZXHb#r...t..T"
+#define EVL_THREAD_STATE_LABELS  "SWDpRUZXHb#r...t..To.."
+
+/* Health monitoring diag codes (via observable or SIGDEBUG). */
+#define EVL_HMDIAG_SIGDEMOTE	1
+#define EVL_HMDIAG_SYSDEMOTE	2
+#define EVL_HMDIAG_EXDEMOTE	3
+#define EVL_HMDIAG_WATCHDOG	4
+#define EVL_HMDIAG_LKDEPEND	5
+#define EVL_HMDIAG_LKIMBALANCE	6
+#define EVL_HMDIAG_LKSLEEP	7
+#define EVL_HMDIAG_STAGEX	8
 
 struct evl_user_window {
 	__u32 state;
diff --git a/include/uapi/evl/types.h b/include/uapi/evl/types.h
index fbc3170a95..975b25db63 100644
--- a/include/uapi/evl/types.h
+++ b/include/uapi/evl/types.h
@@ -44,4 +44,14 @@ struct __evl_itimerspec {
 	struct __evl_timespec it_value;
 };
 
+union evl_value {
+	__s32 val;
+	__s64 lval;
+	void *ptr;
+};
+
+#define evl_intval(__val)	((union evl_value){ .lval = (__val) })
+#define evl_ptrval(__ptr)	((union evl_value){ .ptr = (__ptr) })
+#define evl_nil			evl_intval(0)
+
 #endif /* !_EVL_UAPI_TYPES_H */
diff --git a/kernel/evl/Kconfig b/kernel/evl/Kconfig
index c61a77be71..39dd05b6a6 100644
--- a/kernel/evl/Kconfig
+++ b/kernel/evl/Kconfig
@@ -103,6 +103,21 @@ config EVL_NR_PROXIES
 	be alive concurrently in the system for user-space
 	applications.
 
+config EVL_NR_OBSERVABLES
+	int "Maximum number of observables"
+	range 1 16384
+	default 64
+	help
+
+	This value gives the maximum number of observable elements
+	which can live concurrently in the EVL core. Observables
+	enable the observer design pattern, in which any number of
+	observer threads can be notified of updates to any number of
+	observable subjects, in a loosely coupled fashion. An EVL
+	thread is in and of itself an observable which can be
+	monitored for events; observables attached to threads are not
+	accounted for in this value.
+
 endmenu
 
 menu "Pre-calibrated latency"
@@ -189,13 +204,13 @@ config EVL_WATCHDOG
 	default y
 	help
 	  This option activates a watchdog aimed at detecting runaway
-	  EVL threads. If enabled, the watchdog triggers after a
-	  given period of uninterrupted out-of-band activity has elapsed
-	  without Linux interaction in the meantime.
+	  EVL threads. If enabled, the watchdog triggers after a given
+	  period of uninterrupted out-of-band activity has elapsed
+	  without in-band interaction in the meantime.
 
-	  In such an event, the current thread is moved out the
-	  out-of-band context, receiving a SIGDEBUG signal from the
-	  Linux kernel immediately after.
+	  In such an event, the thread preempted by the watchdog timer
+	  is kicked out the out-of-band context, and immediately
+	  receives a SIGDEBUG signal from the kernel.
 
 	  The timeout value of the watchdog can be set using the
 	  EVL_WATCHDOG_TIMEOUT parameter.
diff --git a/kernel/evl/Makefile b/kernel/evl/Makefile
index 9bdee544b4..474c804014 100644
--- a/kernel/evl/Makefile
+++ b/kernel/evl/Makefile
@@ -9,6 +9,7 @@ evl-y :=		\
 	memory.o	\
 	monitor.o	\
 	mutex.o		\
+	observable.o	\
 	poll.o		\
 	proxy.o		\
 	sem.o		\
diff --git a/kernel/evl/factory.c b/kernel/evl/factory.c
index 5cdbb55d1b..53212c521b 100644
--- a/kernel/evl/factory.c
+++ b/kernel/evl/factory.c
@@ -43,6 +43,7 @@ static struct evl_factory *factories[] = {
 	&evl_poll_factory,
 	&evl_xbuf_factory,
 	&evl_proxy_factory,
+	&evl_observable_factory,
 #ifdef CONFIG_FTRACE
 	&evl_trace_factory,
 #endif
@@ -187,6 +188,7 @@ static void __do_put_element(struct evl_element *e)
 	 * reusing it too early.
 	 */
 	evl_remove_element_device(e);
+
 	/*
 	 * Serialize with evl_open_element().
 	 */
@@ -341,8 +343,8 @@ static int do_element_visibility(struct evl_element *e,
 	struct file *filp;
 	int ret, efd;
 
-	if (EVL_WARN_ON(CORE, !evl_element_is_core(e) && !current->mm))
-		e->clone_flags |= EVL_CLONE_CORE;
+	if (EVL_WARN_ON(CORE, !evl_element_has_coredev(e) && !current->mm))
+		e->clone_flags |= EVL_CLONE_COREDEV;
 
 	/*
 	 * Unlike a private one, a publically visible element exports
@@ -359,7 +361,7 @@ static int do_element_visibility(struct evl_element *e,
 
 	*rdev = MKDEV(0, e->minor);
 
-	if (evl_element_is_core(e))
+	if (evl_element_has_coredev(e))
 		return 0;
 
 	/*
@@ -455,7 +457,7 @@ static int create_element_device(struct evl_element *e,
 	 * cannot fail creating the device anymore. First take a
 	 * reference then install fd (which is a membar).
 	 */
-	if (!evl_element_is_public(e) && !evl_element_is_core(e)) {
+	if (!evl_element_is_public(e) && !evl_element_has_coredev(e)) {
 		e->refs++;
 		fd_install(e->fpriv.efd, e->fpriv.filp);
 	}
@@ -467,7 +469,7 @@ static int create_element_device(struct evl_element *e,
 fail_device:
 	if (evl_element_is_public(e)) {
 		cdev_del(&e->cdev);
-	} else if (!evl_element_is_core(e)) {
+	} else if (!evl_element_has_coredev(e)) {
 		put_unused_fd(e->fpriv.efd);
 		filp_close(e->fpriv.filp, current->files);
 	}
@@ -490,7 +492,7 @@ int evl_create_core_element_device(struct evl_element *e,
 	if (devname == NULL)
 		return PTR_ERR(devname);
 
-	e->clone_flags |= EVL_CLONE_CORE;
+	e->clone_flags |= EVL_CLONE_COREDEV;
 	e->devname = devname;
 
 	return create_element_device(e, fac);
@@ -537,9 +539,6 @@ static long ioctl_clone_device(struct file *filp, unsigned int cmd,
 	if (ret)
 		return -EFAULT;
 
-	if (req.clone_flags & ~EVL_CLONE_MASK)
-		return -EINVAL;
-
 	if (req.name_ptr) {
 		devname = getname(evl_valptr64(req.name_ptr, const char));
 		if (IS_ERR(devname))
@@ -547,8 +546,8 @@ static long ioctl_clone_device(struct file *filp, unsigned int cmd,
 	} else if (req.clone_flags & EVL_CLONE_PUBLIC)
 		return -EINVAL;
 
-	fac = container_of(filp->f_inode->i_cdev, struct evl_factory, cdev);
 	u_attrs = evl_valptr64(req.attrs_ptr, void);
+	fac = container_of(filp->f_inode->i_cdev, struct evl_factory, cdev);
 	e = fac->build(fac, devname ? devname->name : NULL,
 		u_attrs, req.clone_flags, &state_offset);
 	if (IS_ERR(e)) {
@@ -630,9 +629,9 @@ static const struct file_operations clone_fops = {
 #endif
 };
 
-static int index_element_at(struct evl_element *e, fundle_t fundle)
+static int index_element_at(struct evl_index *map,
+			struct evl_element *e, fundle_t fundle)
 {
-	struct evl_index *map = &e->factory->index;
 	struct rb_node **rbp, *parent;
 	struct evl_element *tmp;
 
@@ -656,22 +655,8 @@ static int index_element_at(struct evl_element *e, fundle_t fundle)
 	return 0;
 }
 
-int evl_index_element_at(struct evl_element *e, fundle_t fundle)
-{
-	struct evl_index *map = &e->factory->index;
-	unsigned long flags;
-	int ret;
-
-	raw_spin_lock_irqsave(&map->lock, flags);
-	ret = index_element_at(e, fundle);
-	raw_spin_unlock_irqrestore(&map->lock, flags);
-
-	return ret;
-}
-
-void evl_index_element(struct evl_element *e)
+void evl_index_element(struct evl_index *map, struct evl_element *e)
 {
-	struct evl_index *map = &e->factory->index;
 	fundle_t fundle, guard = 0;
 	unsigned long flags;
 	int ret;
@@ -689,15 +674,14 @@ void evl_index_element(struct evl_element *e)
 		if (!fundle)		/* Exclude EVL_NO_HANDLE */
 			fundle = map->generator = 1;
 
-		ret = index_element_at(e, fundle);
+		ret = index_element_at(map, e, fundle);
 
 		raw_spin_unlock_irqrestore(&map->lock, flags);
 	} while (ret);
 }
 
-void evl_unindex_element(struct evl_element *e)
+void evl_unindex_element(struct evl_index *map, struct evl_element *e)
 {
-	struct evl_index *map = &e->factory->index;
 	unsigned long flags;
 
 	raw_spin_lock_irqsave(&map->lock, flags);
@@ -706,9 +690,8 @@ void evl_unindex_element(struct evl_element *e)
 }
 
 struct evl_element *
-__evl_get_element_by_fundle(struct evl_factory *fac, fundle_t fundle)
+__evl_get_element_by_fundle(struct evl_index *map, fundle_t fundle)
 {
-	struct evl_index *map = &fac->index;
 	struct evl_element *e;
 	unsigned long flags;
 	struct rb_node *rb;
diff --git a/kernel/evl/monitor.c b/kernel/evl/monitor.c
index 6f4674e341..ad04d304c0 100644
--- a/kernel/evl/monitor.c
+++ b/kernel/evl/monitor.c
@@ -106,9 +106,9 @@ void __evl_commit_monitor_ceiling(void)
 	 * curr->u_window has to be valid since curr bears T_USER.  If
 	 * pp_pending is a bad handle, just skip ceiling.
 	 */
-	gate = evl_get_element_by_fundle(&evl_monitor_factory,
-					curr->u_window->pp_pending,
-					struct evl_monitor);
+	gate = evl_get_factory_element_by_fundle(&evl_monitor_factory,
+						curr->u_window->pp_pending,
+						struct evl_monitor);
 	if (gate == NULL)
 		goto out;
 
@@ -770,6 +770,9 @@ monitor_factory_build(struct evl_factory *fac, const char *name,
 	struct evl_clock *clock;
 	int ret;
 
+	if (clone_flags & ~EVL_CLONE_PUBLIC)
+		return ERR_PTR(-EINVAL);
+
 	ret = copy_from_user(&attrs, u_attrs, sizeof(attrs));
 	if (ret)
 		return ERR_PTR(-EFAULT);
@@ -859,7 +862,7 @@ monitor_factory_build(struct evl_factory *fac, const char *name,
 	mon->protocol = attrs.protocol;
 	mon->state = state;
 	*state_offp = evl_shared_offset(state);
-	evl_index_element(&mon->element);
+	evl_index_factory_element(&mon->element);
 
 	return &mon->element;
 
@@ -880,7 +883,7 @@ static void monitor_factory_dispose(struct evl_element *e)
 
 	mon = container_of(e, struct evl_monitor, element);
 
-	evl_unindex_element(&mon->element);
+	evl_unindex_factory_element(&mon->element);
 
 	if (mon->type == EVL_MONITOR_EVENT) {
 		evl_put_clock(mon->wait_queue.clock);
@@ -934,9 +937,9 @@ static ssize_t state_show(struct device *dev,
 	} else {
 		fun = atomic_read(&state->u.gate.owner);
 		if (fun != EVL_NO_HANDLE)
-			owner = evl_get_element_by_fundle(&evl_thread_factory,
-							evl_get_index(fun),
-							struct evl_thread);
+			owner = evl_get_factory_element_by_fundle(&evl_thread_factory,
+						evl_get_index(fun),
+						struct evl_thread);
 		ret = snprintf(buf, PAGE_SIZE, "%d %u %u\n",
 			owner ? evl_get_inband_pid(owner) : -1,
 			state->u.gate.ceiling,
diff --git a/kernel/evl/mutex.c b/kernel/evl/mutex.c
index 59d0ed6585..6c454bffbd 100644
--- a/kernel/evl/mutex.c
+++ b/kernel/evl/mutex.c
@@ -50,7 +50,7 @@ static inline bool enable_inband_switch(struct evl_thread *curr)
 
 	atomic_set(&curr->inband_disable_count, 0);
 	if (curr->state & T_WOLI)
-		evl_signal_thread(curr, SIGDEBUG, SIGDEBUG_MUTEX_IMBALANCE);
+		evl_notify_thread(curr, EVL_HMDIAG_LKIMBALANCE, evl_nil);
 
 	return false;
 }
@@ -396,7 +396,7 @@ static void detect_inband_owner(struct evl_mutex *mutex,
 	} else if (owner->state & T_INBAND) {
 		curr->info |= T_PIALERT;
 		evl_spin_unlock(&curr->rq->lock);
-		evl_signal_thread(curr, SIGDEBUG, SIGDEBUG_MIGRATE_PRIOINV);
+		evl_notify_thread(curr, EVL_HMDIAG_LKDEPEND, evl_nil);
 		return;
 	}
 
@@ -404,39 +404,38 @@ static void detect_inband_owner(struct evl_mutex *mutex,
 }
 
 /*
- * Detect when a thread is about to switch in-band while holding a
- * mutex which is causing an active PI or PP boost. Since this would
- * cause a priority inversion, any thread waiting for this mutex
- * bearing the T_WOLI bit receives a SIGDEBUG notification in this
- * case.
+ * Detect when current is about to switch in-band while holding a
+ * mutex which is causing an active PI or PP boost. Since such a
+ * dependency on in-band would cause a priority inversion for the
+ * waiter(s), the latter is sent a HM notification if T_WOLI is set.
  */
-void evl_detect_boost_drop(struct evl_thread *owner)
+void evl_detect_boost_drop(void)
 {
+	struct evl_thread *curr = evl_current();
 	struct evl_thread *waiter;
 	struct evl_mutex *mutex;
 	unsigned long flags;
 
-	evl_spin_lock_irqsave(&owner->lock, flags);
+	evl_spin_lock_irqsave(&curr->lock, flags);
 
 	/*
 	 * Iterate over waiters of each mutex we got boosted for due
 	 * to PI/PP.
 	 */
-	for_each_evl_booster(mutex, owner) {
+	for_each_evl_booster(mutex, curr) {
 		evl_spin_lock(&mutex->lock);
 		for_each_evl_mutex_waiter(waiter, mutex) {
-			if (waiter->state & T_WOLI) {
-				evl_spin_lock(&waiter->rq->lock);
-				waiter->info |= T_PIALERT;
-				evl_spin_unlock(&waiter->rq->lock);
-				evl_signal_thread(waiter, SIGDEBUG,
-						SIGDEBUG_MIGRATE_PRIOINV);
-			}
+			if (!(waiter->state & T_WOLI))
+				continue;
+			evl_spin_lock(&waiter->rq->lock);
+			waiter->info |= T_PIALERT;
+			evl_spin_unlock(&waiter->rq->lock);
+			evl_notify_thread(waiter, EVL_HMDIAG_LKDEPEND, evl_nil);
 		}
 		evl_spin_unlock(&mutex->lock);
 	}
 
-	evl_spin_unlock_irqrestore(&owner->lock, flags);
+	evl_spin_unlock_irqrestore(&curr->lock, flags);
 }
 
 void __evl_init_mutex(struct evl_mutex *mutex,
@@ -675,7 +674,7 @@ int evl_lock_mutex_timeout(struct evl_mutex *mutex, ktime_t timeout,
 		h = oldh;
 	} while (!fast_mutex_is_claimed(h));
 
-	owner = evl_get_element_by_fundle(&evl_thread_factory,
+	owner = evl_get_factory_element_by_fundle(&evl_thread_factory,
 					evl_get_index(h),
 					struct evl_thread);
 	/*
@@ -711,9 +710,10 @@ int evl_lock_mutex_timeout(struct evl_mutex *mutex, ktime_t timeout,
 		track_owner(mutex, owner);
 	else
 		/*
-		 * evl_get_element_by_fundle() got us an extraneous
-		 * reference on @owner which an earlier call to
-		 * track_owner() already obtained, drop the former.
+		 * evl_get_factory_element_by_fundle() got us an
+		 * extraneous reference on @owner which an earlier
+		 * call to track_owner() already obtained, drop the
+		 * former.
 		 */
 		evl_put_element(&owner->element);
 
diff --git a/kernel/evl/observable.c b/kernel/evl/observable.c
new file mode 100644
index 0000000000..9cf53c774c
--- /dev/null
+++ b/kernel/evl/observable.c
@@ -0,0 +1,1147 @@
+/*
+ * SPDX-License-Identifier: GPL-2.0
+ *
+ * Copyright (C) 2020 Philippe Gerum  <rpm@xenomai.org>
+ */
+
+#include <linux/types.h>
+#include <linux/slab.h>
+#include <linux/rbtree.h>
+#include <linux/poll.h>
+#include <linux/rbtree.h>
+#include <evl/list.h>
+#include <evl/sched.h>
+#include <evl/thread.h>
+#include <evl/clock.h>
+#include <evl/observable.h>
+#include <evl/factory.h>
+
+/* Co-exists with EVL_NOTIFY_MASK bits. */
+#define EVL_NOTIFY_INITIAL	(1 << 31)
+
+#define EVL_OBSERVABLE_CLONE_FLAGS	\
+	(EVL_CLONE_PUBLIC|EVL_CLONE_OBSERVABLE|EVL_CLONE_MASTER)
+
+/*
+ * We want any kind of threads to be able to subscribe to an
+ * observable including non-EVL ones, evl_subscriber holds the
+ * information we need for this. This descriptor is accessed directly
+ * from the task_struct via the oob thread state.
+ */
+struct evl_subscriber {
+  	struct rb_root subscriptions; /* struct evl_observer */
+	evl_spinlock_t lock;
+};
+
+struct evl_notification_record {
+	u32 tag;
+	u32 serial;
+	pid_t issuer;
+	union evl_value event;
+	ktime_t date;
+	struct list_head next;	/* in evl_observers.(free|pending)_list */
+};
+
+struct evl_observer {
+	size_t backlog_size;
+	struct list_head free_list;	/* evl_notification_record */
+	struct list_head pending_list;	/* evl_notification_record */
+	struct list_head next;		/* in evl_observable.observers */
+	fundle_t fundle;		/* fundle of observable */
+	struct rb_node rb;		/* in evl_subscriber.subscriptions */
+	int flags;			/* EVL_NOTIFY_xx */
+	int refs;			/* notification vs unsubscription */
+	struct evl_notice last_notice;
+	struct evl_notification_record backlog[0];
+};
+
+static const struct file_operations observable_fops;
+
+static inline
+int index_observer(struct rb_root *root, struct evl_observer *observer)
+{
+	struct rb_node **rbp, *parent = NULL;
+	struct evl_observer *tmp;
+
+	rbp = &root->rb_node;
+	while (*rbp) {
+		tmp = rb_entry(*rbp, struct evl_observer, rb);
+		parent = *rbp;
+		if (observer->fundle < tmp->fundle)
+			rbp = &(*rbp)->rb_left;
+		else if (observer->fundle > tmp->fundle)
+			rbp = &(*rbp)->rb_right;
+		else
+			return -EEXIST;
+	}
+
+	rb_link_node(&observer->rb, parent, rbp);
+	rb_insert_color(&observer->rb, root);
+
+	return 0;
+}
+
+static struct evl_observer *
+find_observer(struct rb_root *root, fundle_t fundle)
+{
+	struct evl_observer *observer;
+	struct rb_node *rb;
+
+	rb = root->rb_node;
+	while (rb) {
+		observer = rb_entry(rb, struct evl_observer, rb);
+		if (fundle < observer->fundle)
+			rb = rb->rb_left;
+		else if (fundle > observer->fundle)
+			rb = rb->rb_right;
+		else
+			return observer;
+	}
+
+	return NULL;
+}
+
+/*
+ * Threads can only subscribe/unsubscribe to/from observables for
+ * themselves, which guarantees that no observer can go away while a
+ * thread is accessing it from [oob_]read()/write() operations.  An
+ * observer only knows about the observable's fundle, it does not
+ * maintain any memory reference to the former: we want lose coupling
+ * in this direction so that observables can go away while subscribers
+ * still target them.
+ */
+static int add_subscription(struct evl_observable *observable,
+			unsigned int backlog_count, int op_flags) /* in-band */
+{
+	struct evl_subscriber *sbr = evl_get_subscriber();
+	struct evl_notification_record *nf;
+	struct evl_observer *observer;
+	size_t backlog_size;
+	unsigned long flags;
+	unsigned int n;
+	int ret;
+
+	inband_context_only();
+
+	if (backlog_count == 0)
+		return -EINVAL;
+
+	if (op_flags & ~EVL_NOTIFY_MASK)
+		return -EINVAL;
+
+	if (sbr == NULL) {
+		sbr = kzalloc(sizeof(*sbr), GFP_KERNEL);
+		if (sbr == NULL)
+			return -ENOMEM;
+		evl_spin_lock_init(&sbr->lock);
+		sbr->subscriptions = RB_ROOT;
+		evl_set_subscriber(sbr);
+	}
+
+	/*
+	 * We don't need to guard against racing with in-band file
+	 * release since the caller must be running in-band, so the
+	 * VFS already ensures consistency between release vs other
+	 * in-band I/O ops. We do need a reference on the underlying
+	 * element though.
+	 */
+	evl_get_element(&observable->element);
+
+	/*
+	 * The observer descriptor and its backlog are adjacent in
+	 * memory. The backlog contains notification records, align on
+	 * this element size.
+	 */
+	backlog_size = backlog_count * sizeof(*nf);
+	observer = kzalloc(sizeof(*observer) + backlog_size, GFP_KERNEL);
+	if (observer == NULL) {
+		ret = -ENOMEM;
+		goto fail_alloc;
+	}
+
+	observer->fundle = fundle_of(observable);
+	observer->backlog_size = backlog_size;
+	observer->flags = op_flags;
+	observer->refs = 1;
+	INIT_LIST_HEAD(&observer->free_list);
+	INIT_LIST_HEAD(&observer->pending_list);
+
+	if (op_flags & EVL_NOTIFY_ONCHANGE)
+		observer->flags |= EVL_NOTIFY_INITIAL;
+
+	evl_spin_lock_irqsave(&sbr->lock, flags);
+	ret = index_observer(&sbr->subscriptions, observer);
+	evl_spin_unlock_irqrestore(&sbr->lock, flags);
+	if (ret)
+		goto dup_subscription;
+
+	/* Build the free list of notification records. */
+	for (nf = observer->backlog, n = 0; n < backlog_count; n++, nf++)
+		list_add(&nf->next, &observer->free_list);
+
+	/*
+	 * Make the new observer visible to the observable. We need
+	 * the observer and writability tracking to be atomically
+	 * updated, so nested locking is required. Observers are
+	 * linked to the observable's list by order of subscription
+	 * (FIFO) in case users make such assumption for master mode
+	 * (which undergoes round-robin).
+	 */
+	evl_spin_lock_irqsave(&observable->lock, flags);
+	list_add_tail(&observer->next, &observable->observers);
+	evl_spin_lock(&observable->oob_wait.lock);
+	observable->writable_observers++;
+	evl_spin_unlock(&observable->oob_wait.lock);
+	evl_spin_unlock_irqrestore(&observable->lock, flags);
+	evl_signal_poll_events(&observable->poll_head, POLLOUT|POLLWRNORM);
+
+	evl_put_element(&observable->element);
+
+	return 0;
+
+dup_subscription:
+	kfree(observer);
+fail_alloc:
+	evl_put_element(&observable->element);
+
+	return ret;
+}
+
+/* observable->oob_wait.lock held, irqs off */
+static void decrease_writability(struct evl_observable *observable)
+{
+	observable->writable_observers--;
+	EVL_WARN_ON(CORE, observable->writable_observers < 0);
+}
+
+/* observable->lock held, irqs off */
+static void detach_observer_locked(struct evl_observer *observer,
+		struct evl_observable *observable) /* irqs off */
+{
+	list_del(&observer->next);
+
+	evl_spin_lock(&observable->oob_wait.lock);
+
+	if (!list_empty(&observer->free_list))
+		decrease_writability(observable);
+
+	evl_spin_unlock(&observable->oob_wait.lock);
+}
+
+static inline void get_observer(struct evl_observer *observer)
+{
+	observer->refs++;
+}
+
+static inline bool put_observer(struct evl_observer *observer)
+{
+	int new_refs = --observer->refs;
+
+	EVL_WARN_ON(CORE, new_refs < 0);
+
+	return new_refs == 0;
+}
+
+static bool detach_observer(struct evl_observer *observer,
+			struct evl_observable *observable)
+{
+	bool dropped = false;
+	unsigned long flags;
+
+	evl_spin_lock_irqsave(&observable->lock, flags);
+
+	/*
+	 * If the refcount does not drop to zero, push_notification()
+	 * will flush the observer later on when it is safe.
+	 */
+	if (likely(put_observer(observer))) {
+		detach_observer_locked(observer, observable);
+		dropped = true;
+	}
+
+	evl_spin_unlock_irqrestore(&observable->lock, flags);
+
+	return dropped;
+}
+
+static int cancel_subscription(struct evl_observable *observable) /* in-band */
+{
+	struct evl_subscriber *sbr = evl_get_subscriber();
+	struct evl_observer *observer;
+	unsigned long flags;
+	bool dropped;
+
+	inband_context_only();
+
+	if (sbr == NULL)
+		return -ENOENT;
+
+	evl_get_element(&observable->element);
+
+	observer = find_observer(&sbr->subscriptions,
+				fundle_of(observable));
+	if (observer == NULL) {
+		evl_put_element(&observable->element);
+		return -ENOENT;
+	}
+
+	/*
+	 * Remove observer from the subscription index. Once done,
+	 * this observer escapes evl_drop_subscriptions() in case the
+	 * caller exits afterwards.
+	 */
+	evl_spin_lock_irqsave(&sbr->lock, flags);
+	rb_erase(&observer->rb, &sbr->subscriptions);
+	evl_spin_unlock_irqrestore(&sbr->lock, flags);
+
+	/*
+	 * Remove observer from observable->observers. Subscription
+	 * cannot be stale since observable is still around (meaning
+	 * observable_release() did not run).
+	 */
+	dropped = detach_observer(observer, observable);
+
+	evl_put_element(&observable->element);
+
+	if (dropped)
+		kfree(observer);
+
+	return 0;
+}
+
+/*
+ * Current is wrapping up, drop the associated subscriber information
+ * if any, no need to lock subscriptions.
+ */
+void evl_drop_subscriptions(struct evl_subscriber *sbr)
+{
+	struct evl_observable *observable;
+	struct evl_observer *observer;
+	bool dropped = true;
+	struct rb_node *rb;
+
+	if (sbr == NULL)
+		return;
+
+	while ((rb = sbr->subscriptions.rb_node) != NULL) {
+		observer = rb_entry(rb, struct evl_observer, rb);
+		rb_erase(rb, &sbr->subscriptions);
+		/* Some subscriptions might be stale, check this. */
+		if (likely(!list_empty(&observer->next))) {
+			observable = evl_get_factory_element_by_fundle(
+				&evl_observable_factory,
+				observer->fundle,
+				struct evl_observable);
+			if (!EVL_WARN_ON(CORE, observable == NULL)) {
+				dropped = detach_observer(observer, observable);
+				evl_put_element(&observable->element);
+			} else {
+				/* Something is going seriously wrong. */
+				dropped = false;
+			}
+		}
+		if (dropped)
+			kfree(observer);
+	}
+
+	kfree(sbr);
+}
+
+static void wake_oob_threads(struct evl_observable *observable)
+{
+	unsigned long flags;
+
+	evl_spin_lock_irqsave(&observable->oob_wait.lock, flags);
+
+	if (evl_wait_active(&observable->oob_wait))
+		evl_flush_wait_locked(&observable->oob_wait, 0);
+
+	evl_spin_unlock_irqrestore(&observable->oob_wait.lock, flags);
+
+	evl_schedule();
+}
+
+static void wake_inband_threads(struct evl_observable *observable) /* in-band */
+{
+	wake_up_all(&observable->inband_wait);
+}
+
+static void inband_wake_irqwork(struct irq_work *work)
+{
+	struct evl_observable *observable;
+
+	observable = container_of(work, struct evl_observable, wake_irqwork);
+	wake_inband_threads(observable);
+	evl_put_element(&observable->element);
+}
+
+void evl_flush_observable(struct evl_observable *observable)
+{
+	struct evl_observer *observer, *tmp;
+	unsigned long flags;
+
+	/*
+	 * We may be called from do_cleanup_current() in which case
+	 * some events might still being pushed to the observable via
+	 * the thread file descriptor, so locking is required.
+	 */
+	evl_spin_lock_irqsave(&observable->oob_wait.lock, flags);
+
+	list_for_each_entry_safe(observer, tmp, &observable->observers, next) {
+		list_del_init(&observer->next);
+		if (!list_empty(&observer->free_list))
+			decrease_writability(observable);
+	}
+
+	evl_spin_unlock_irqrestore(&observable->oob_wait.lock, flags);
+
+	wake_oob_threads(observable);
+}
+
+static int observable_release(struct inode *inode, struct file *filp)
+{
+	struct evl_observable *observable = element_of(filp, struct evl_observable);
+
+	/*
+	 * Unbind observers, then wake up so that oob waiters can
+	 * figure out that we forced unsubscribed them. Those
+	 * subscriptions are now stale. There is no point in
+	 * attempting to wake up in-band waiters, there cannot be any
+	 * since no regular read()/write() request could be ongoing on
+	 * this file since the VFS wants us to release it.
+	 */
+	evl_flush_observable(observable);
+
+	/*
+	 * Wait for any ongoing oob request to have completed before
+	 * allowing the VFS to fully dismantle filp.
+	 */
+	return evl_release_element(inode, filp);
+}
+
+static void do_flush_work(struct evl_observable *observable)
+{
+	struct evl_observer *observer;
+	unsigned long flags;
+
+	/*
+	 * Flush the observers pending deletion. This should happen
+	 * only rarely, in case push_notification() raced with
+	 * cancel_subscription() or evl_drop_subscriptions().
+	 */
+	evl_spin_lock_irqsave(&observable->lock, flags);
+
+	while (!list_empty(&observable->flush_list)) {
+		observer = list_get_entry(&observable->flush_list,
+					struct evl_observer, next);
+		evl_spin_unlock_irqrestore(&observable->lock, flags);
+		kfree(observer);
+		evl_spin_lock_irqsave(&observable->lock, flags);
+	}
+
+	evl_spin_unlock_irqrestore(&observable->lock, flags);
+}
+
+static void inband_flush_irqwork(struct irq_work *work)
+{
+	struct evl_observable *observable;
+
+	observable = container_of(work, struct evl_observable, flush_irqwork);
+	do_flush_work(observable);
+	evl_put_element(&observable->element);
+}
+
+static bool notify_one_observer(struct evl_observable *observable,
+			struct evl_observer *observer,
+			const struct evl_notification_record *tmpl_nfr)
+{
+	struct evl_notification_record *nfr;
+	struct evl_notice last_notice;
+	unsigned long flags;
+
+	evl_spin_lock_irqsave(&observable->oob_wait.lock, flags);
+
+	if (observer->flags & EVL_NOTIFY_ONCHANGE) {
+		last_notice = observer->last_notice;
+		observer->last_notice.tag = tmpl_nfr->tag;
+		observer->last_notice.event = tmpl_nfr->event;
+		if (unlikely(observer->flags & EVL_NOTIFY_INITIAL))
+			observer->flags &= ~EVL_NOTIFY_INITIAL;
+		else if (tmpl_nfr->tag == last_notice.tag &&
+			tmpl_nfr->event.lval == last_notice.event.lval)
+			goto done;
+	}
+
+	if (list_empty(&observer->free_list)) {
+		evl_spin_unlock_irqrestore(&observable->oob_wait.lock, flags);
+		return false;
+	}
+
+	nfr = list_get_entry(&observer->free_list,
+			struct evl_notification_record, next);
+
+	*nfr = *tmpl_nfr;
+	list_add_tail(&nfr->next, &observer->pending_list);
+
+	if (list_empty(&observer->free_list))
+		decrease_writability(observable);
+done:
+	evl_spin_unlock_irqrestore(&observable->oob_wait.lock, flags);
+
+	return true;
+}
+
+static bool notify_single(struct evl_observable *observable,
+			struct evl_notification_record *nfr,
+			ssize_t *len_r)
+{
+	struct evl_observer *observer;
+	bool do_flush = false;
+	unsigned long flags;
+
+	evl_spin_lock_irqsave(&observable->lock, flags);
+
+	if (list_empty(&observable->observers))
+		goto out;
+
+	nfr->serial = observable->serial_counter++;
+
+	observer = list_first_entry(&observable->observers,
+				struct evl_observer, next);
+	get_observer(observer);
+
+	/*
+	 * Use round-robin as a naive load-balancing strategy among
+	 * multiple observers forming a pool of worker threads.
+	 */
+	if (!list_is_singular(&observable->observers)) {
+		list_del(&observer->next);
+		list_add_tail(&observer->next, &observable->observers);
+	}
+
+	evl_spin_unlock_irqrestore(&observable->lock, flags);
+
+	if (notify_one_observer(observable, observer, nfr))
+		*len_r += sizeof(struct evl_notice);
+
+	evl_spin_lock_irqsave(&observable->lock, flags);
+
+	if (unlikely(put_observer(observer))) {
+		detach_observer_locked(observer, observable);
+		list_add(&observer->next, &observable->flush_list);
+		do_flush = true;
+	}
+out:
+	evl_spin_unlock_irqrestore(&observable->lock, flags);
+
+	return do_flush;
+}
+
+static bool notify_all(struct evl_observable *observable,
+			struct evl_notification_record *nfr,
+			ssize_t *len_r)
+{
+	struct list_head *pos, *nextpos;
+	struct evl_observer *observer;
+	unsigned long flags;
+	bool do_flush;
+
+	evl_spin_lock_irqsave(&observable->lock, flags);
+
+	if (list_empty(&observable->observers)) {
+		do_flush = false;
+		goto out;
+	}
+
+	nfr->serial = observable->serial_counter++;
+
+	observer = list_first_entry(&observable->observers,
+				struct evl_observer, next);
+	get_observer(observer);
+
+	/*
+	 * Prevent cancel_subscription() and evl_drop_subscriptions()
+	 * from dropping the observers we are about to push the
+	 * notification to, without locking the observable all over
+	 * the list walk. The trick is to hold a reference both on the
+	 * current and next items during the traversal. Observers
+	 * which might be added to the list concurrently by
+	 * add_subscription() would not be picked during the
+	 * traversal, which is still logically correct.
+	 */
+	list_for_each_safe(pos, nextpos, &observable->observers) {
+
+		if (likely(nextpos != &observable->observers)) {
+			observer = list_entry(nextpos, struct evl_observer, next);
+			get_observer(observer);
+		}
+
+		observer = list_entry(pos, struct evl_observer, next);
+
+		evl_spin_unlock_irqrestore(&observable->lock, flags);
+
+		if (notify_one_observer(observable, observer, nfr))
+			*len_r += sizeof(struct evl_notice);
+
+		evl_spin_lock_irqsave(&observable->lock, flags);
+
+		if (unlikely(put_observer(observer))) {
+			detach_observer_locked(observer, observable);
+			list_add(&observer->next, &observable->flush_list);
+		}
+	}
+
+	do_flush = !list_empty(&observable->flush_list);
+out:
+	evl_spin_unlock_irqrestore(&observable->lock, flags);
+
+	return do_flush;
+}
+
+static bool is_pool_master(struct evl_observable *observable)
+{
+	return !!(observable->element.clone_flags & EVL_CLONE_MASTER);
+}
+
+static bool push_notification(struct evl_observable *observable,
+			const struct evl_notice *ntc,
+			ktime_t date)
+{
+	struct evl_notification_record nfr;
+	ssize_t len = 0;
+	bool do_flush;
+
+	/*
+	 * Prep a template notification record so that multiple
+	 * observers receive exactly the same notification data for
+	 * any single change, timestamp included.
+	 */
+	nfr.tag = ntc->tag;
+	nfr.event = ntc->event;
+	nfr.issuer = ntc->tag >= EVL_NOTICE_USER ? task_pid_nr(current) : 0;
+	nfr.date = date;
+
+	/*
+	 * Feed one observer/worker if master, or broadcast to all
+	 * observers.
+	 */
+	if (is_pool_master(observable))
+		do_flush = notify_single(observable, &nfr, &len);
+	else
+		do_flush = notify_all(observable, &nfr, &len);
+
+	if (unlikely(do_flush)) {
+		if (running_inband()) {
+			do_flush_work(observable);
+		} else {
+			evl_get_element(&observable->element);
+			if (!irq_work_queue(&observable->flush_irqwork))
+				evl_put_element(&observable->element);
+		}
+	}
+
+	return len > 0;  /* True if some subscribers could receive. */
+}
+
+static void wake_up_observers(struct evl_observable *observable)
+{
+	wake_oob_threads(observable);
+	evl_signal_poll_events(&observable->poll_head, POLLIN|POLLRDNORM);
+
+	/*
+	 * If running oob, we need to go through the wake_irqwork
+	 * trampoline for waking up in-band waiters.
+	 */
+	if (running_inband()) {
+		wake_inband_threads(observable);
+	} else {
+		evl_get_element(&observable->element);
+		if (!irq_work_queue(&observable->wake_irqwork))
+			evl_put_element(&observable->element);
+	}
+}
+
+bool evl_send_observable(struct evl_observable *observable, int tag,
+			union evl_value details)
+{
+	struct evl_notice ntc;
+	ktime_t now;
+
+	ntc.tag = tag;
+	ntc.event = details;
+	now = evl_ktime_monotonic();
+
+	if (push_notification(observable, &ntc, now)) {
+		wake_up_observers(observable);
+		return true;
+	}
+
+	return false;
+}
+EXPORT_SYMBOL_GPL(evl_send_observable);
+
+ssize_t evl_write_observable(struct evl_observable *observable,
+			const char __user *u_buf, size_t count)
+{
+	ssize_t len = 0, xlen = 0;
+	struct evl_notice ntc;
+	ktime_t now;
+	int ret = 0;
+
+	if (!evl_element_is_observable(&observable->element))
+		return -ENXIO;
+
+	if (count == 0)
+		return 0;
+
+	if (count % sizeof(struct evl_notice))
+		return -EINVAL;	/* No partial write. */
+
+	now = evl_ktime_monotonic();
+
+	while (xlen < count) {
+		if (raw_copy_from_user(&ntc, u_buf, sizeof(ntc))) {
+			ret = -EFAULT;
+			break;
+		}
+
+		if (ntc.tag < EVL_NOTICE_USER) {
+			ret = -EINVAL;
+			break;
+		}
+
+		if (push_notification(observable, &ntc, now))
+			len += sizeof(ntc);
+
+		xlen += sizeof(ntc);
+		u_buf += sizeof(ntc);
+	}
+
+	/* If some notification was delivered, wake observers up. */
+	if (len > 0)
+		wake_up_observers(observable);
+
+	return ret ?: len;
+}
+
+static struct evl_notification_record *
+pull_from_oob(struct evl_observable *observable,
+	struct evl_observer *observer,
+	bool wait)
+{
+	struct evl_notification_record *nfr;
+	unsigned long flags;
+	int ret;
+
+	evl_spin_lock_irqsave(&observable->oob_wait.lock, flags);
+
+	/*
+	 * observable->wait.lock guards the pending and free
+	 * notification lists of all observers subscribed to it.
+	 */
+	for (;;) {
+		if (list_empty(&observer->next)) {
+			/* Unsubscribed by observable_release(). */
+			nfr = ERR_PTR(-EBADF);
+			goto out;
+		}
+		if (!list_empty(&observer->pending_list))
+			break;
+		if (!wait) {
+			nfr = ERR_PTR(-EWOULDBLOCK);
+			goto out;
+		}
+		evl_add_wait_queue(&observable->oob_wait, EVL_INFINITE, EVL_REL);
+		evl_spin_unlock_irqrestore(&observable->oob_wait.lock, flags);
+		ret = evl_wait_schedule(&observable->oob_wait);
+		if (ret)
+			return ERR_PTR(ret);
+		evl_spin_lock_irqsave(&observable->oob_wait.lock, flags);
+	}
+
+	nfr = list_get_entry(&observer->pending_list,
+			struct evl_notification_record, next);
+out:
+	evl_spin_unlock_irqrestore(&observable->oob_wait.lock, flags);
+
+	return nfr;
+}
+
+static struct evl_notification_record *
+pull_from_inband(struct evl_observable *observable,
+		struct evl_observer *observer,
+		bool wait)
+{
+	struct evl_notification_record *nfr = NULL;
+	unsigned long ib_flags, oob_flags;
+	struct wait_queue_entry wq_entry;
+	int ret = 0;
+
+	init_wait_entry(&wq_entry, 0);
+
+	/*
+	 * In-band lock first, oob next. See stax implementation for
+	 * an explanation.
+	 */
+	spin_lock_irqsave(&observable->inband_wait.lock, ib_flags);
+	evl_spin_lock_irqsave(&observable->oob_wait.lock, oob_flags);
+
+	for (;;) {
+		/*
+		 * No need to check for observer->next linkage when
+		 * handling read(): observable_release() did not run
+		 * for the underlying file, obviously.
+		 */
+		if (list_empty(&wq_entry.entry))
+			__add_wait_queue(&observable->inband_wait, &wq_entry);
+
+		if (!list_empty(&observer->pending_list)) {
+			nfr = list_get_entry(&observer->pending_list,
+					struct evl_notification_record, next);
+			break;
+		}
+		if (!wait) {
+			ret = -EWOULDBLOCK;
+			break;
+		}
+		set_current_state(TASK_INTERRUPTIBLE);
+		evl_spin_unlock_irqrestore(&observable->oob_wait.lock, oob_flags);
+		spin_unlock_irqrestore(&observable->inband_wait.lock, ib_flags);
+		schedule();
+		spin_lock_irqsave(&observable->inband_wait.lock, ib_flags);
+		evl_spin_lock_irqsave(&observable->oob_wait.lock, oob_flags);
+
+		if (signal_pending(current)) {
+			ret = -ERESTARTSYS;
+			break;
+		}
+	}
+
+	list_del(&wq_entry.entry);
+
+	evl_spin_unlock_irqrestore(&observable->oob_wait.lock, oob_flags);
+	spin_unlock_irqrestore(&observable->inband_wait.lock, ib_flags);
+
+	return ret ? ERR_PTR(ret) : nfr;
+}
+
+static int pull_notification(struct evl_observable *observable,
+			struct evl_observer *observer,
+			char __user *u_buf,
+			bool wait)
+{
+	struct evl_notification_record *nfr;
+	struct __evl_notification nf;
+	bool sigpoll = false;
+	unsigned long flags;
+	int ret;
+
+	if (running_inband())
+		nfr = pull_from_inband(observable, observer, wait);
+	else
+		nfr = pull_from_oob(observable, observer, wait);
+
+	if (IS_ERR(nfr))
+		return PTR_ERR(nfr);
+
+	nf.tag = nfr->tag;
+	nf.event = nfr->event;
+	nf.issuer = nfr->issuer;
+	nf.serial = nfr->serial;
+	nf.date = ktime_to_u_timespec(nfr->date);
+
+	/*
+	 * Do not risk losing notifications because of a broken
+	 * userland; re-post them at front in case we fail copying the
+	 * information back. A more recent record might have slipped
+	 * in since we dropped the lock, but monotonic timestamps can
+	 * still be used to sort out order of arrival.
+	 */
+	ret = raw_copy_to_user(u_buf, &nf, sizeof(nf));
+
+	evl_spin_lock_irqsave(&observable->oob_wait.lock, flags);
+
+	if (ret) {
+		list_add(&nfr->next, &observer->pending_list);
+		ret = -EFAULT;
+	} else {
+		if (list_empty(&observer->free_list)) {
+			observable->writable_observers++;
+			sigpoll = true;
+		}
+		list_add(&nfr->next, &observer->free_list);
+	}
+
+	evl_spin_unlock_irqrestore(&observable->oob_wait.lock, flags);
+
+	if (unlikely(sigpoll))
+		evl_signal_poll_events(&observable->poll_head,
+				POLLOUT|POLLWRNORM);
+
+	return ret;
+}
+
+ssize_t evl_read_observable(struct evl_observable *observable,
+			char __user *u_buf, size_t count, bool wait)
+{
+	struct evl_subscriber *sbr = evl_get_subscriber();
+	struct evl_observer *observer;
+	ssize_t len = 0;
+	int ret = 0;
+
+	/*
+	 * The caller must have a valid subscription for reading an
+	 * observable.
+	 */
+	if (sbr == NULL)
+		return -ENXIO;
+
+	if (!evl_element_is_observable(&observable->element))
+		return -ENXIO;
+
+	/*
+	 * Invariant: the caller owns observer (if found), nobody else
+	 * may free it. This is _not_ true for push_notification()
+	 * walking the list of observers which may belong to different
+	 * threads.
+	 */
+	observer = find_observer(&sbr->subscriptions, fundle_of(observable));
+	if (observer == NULL)
+		return -ENXIO;	/* Not subscribed. */
+
+	if (count == 0)
+		return 0;
+
+	if (count % sizeof(struct __evl_notification))
+		return -EINVAL;	/* No partial read. */
+	/*
+	 * Return as many available notifications we can, waiting for
+	 * the first one if necessary.
+	 */
+	while (len < count) {
+		ret = pull_notification(observable, observer, u_buf, wait);
+		if (ret)
+			break;
+		wait = false;
+		u_buf += sizeof(struct __evl_notification);
+		len += sizeof(struct __evl_notification);
+	}
+
+	return len ?: ret;
+}
+
+static __poll_t poll_observable(struct evl_observable *observable)
+{
+	struct evl_subscriber *sbr = evl_get_subscriber();
+	struct evl_observer *observer;
+	unsigned long flags;
+	__poll_t ret = 0;
+
+	if (sbr == NULL)
+		return POLLERR;
+
+	if (!evl_element_is_observable(&observable->element))
+		return POLLERR;
+
+	observer = find_observer(&sbr->subscriptions, fundle_of(observable));
+
+	evl_spin_lock_irqsave(&observable->oob_wait.lock, flags);
+
+	/* Only subscribers can inquire about readability. */
+	if (observer && !list_empty(&observer->pending_list))
+		ret = POLLIN|POLLRDNORM;
+
+	if (observable->writable_observers > 0)
+		ret |= POLLOUT|POLLWRNORM;
+
+	evl_spin_unlock_irqrestore(&observable->oob_wait.lock, flags);
+
+	return ret;
+}
+
+__poll_t evl_oob_poll_observable(struct evl_observable *observable,
+				struct oob_poll_wait *wait)
+{
+	evl_poll_watch(&observable->poll_head, wait, NULL);
+
+	return poll_observable(observable);
+}
+
+__poll_t evl_poll_observable(struct evl_observable *observable,
+			struct file *filp, poll_table *pt)
+{
+	poll_wait(filp, &observable->inband_wait, pt);
+
+	return poll_observable(observable);
+}
+
+long evl_ioctl_observable(struct evl_observable *observable, unsigned int cmd,
+			unsigned long arg)
+{
+	struct evl_subscription sub, *u_sub;
+	int ret;
+
+	switch (cmd) {
+	case EVL_OBSIOC_SUBSCRIBE:
+		/*
+		 * Add a subscription to @observable for the current
+		 * thread, which may not be attached.
+		 */
+		u_sub = (typeof(u_sub))arg;
+		ret = raw_copy_from_user(&sub, u_sub, sizeof(sub));
+		if (ret)
+			return -EFAULT;
+		ret = add_subscription(observable,
+				sub.backlog_count, sub.flags);
+		break;
+	case EVL_OBSIOC_UNSUBSCRIBE:
+		/* Likewise for unsubscribing from @observable. */
+		ret = cancel_subscription(observable);
+		break;
+	default:
+		ret = -ENOTTY;
+	}
+
+	evl_schedule();
+
+	return ret;
+}
+
+static ssize_t observable_oob_write(struct file *filp,
+				const char __user *u_buf, size_t count)
+{
+	struct evl_observable *observable = element_of(filp, struct evl_observable);
+
+	return evl_write_observable(observable, u_buf, count);
+}
+
+static ssize_t observable_oob_read(struct file *filp,
+				char __user *u_buf, size_t count)
+{
+	struct evl_observable *observable = element_of(filp, struct evl_observable);
+
+	return evl_read_observable(observable, u_buf, count,
+				!(filp->f_flags & O_NONBLOCK));
+}
+
+static __poll_t observable_oob_poll(struct file *filp,
+			struct oob_poll_wait *wait)
+{
+	struct evl_observable *observable = element_of(filp, struct evl_observable);
+
+	return evl_oob_poll_observable(observable, wait);
+}
+
+static ssize_t observable_write(struct file *filp, const char __user *u_buf,
+				size_t count, loff_t *ppos)
+{
+	struct evl_observable *observable = element_of(filp, struct evl_observable);
+
+	return evl_write_observable(observable, u_buf, count);
+}
+
+static ssize_t observable_read(struct file *filp, char __user *u_buf,
+			size_t count, loff_t *ppos)
+{
+	struct evl_observable *observable = element_of(filp, struct evl_observable);
+
+	return evl_read_observable(observable, u_buf, count,
+				!(filp->f_flags & O_NONBLOCK));
+}
+
+static __poll_t observable_poll(struct file *filp, poll_table *pt)
+{
+	struct evl_observable *observable = element_of(filp, struct evl_observable);
+
+	return evl_poll_observable(observable, filp, pt);
+}
+
+static long observable_ioctl(struct file *filp, unsigned int cmd,
+			unsigned long arg)
+{
+	struct evl_observable *observable = element_of(filp, struct evl_observable);
+
+	return evl_ioctl_observable(observable, cmd, arg);
+}
+
+static const struct file_operations observable_fops = {
+	.open		= evl_open_element,
+	.release	= observable_release,
+	.oob_write	= observable_oob_write,
+	.oob_read	= observable_oob_read,
+	.oob_poll	= observable_oob_poll,
+	.write		= observable_write,
+	.read		= observable_read,
+	.poll		= observable_poll,
+	.unlocked_ioctl	= observable_ioctl,
+#ifdef CONFIG_COMPAT
+	.compat_ioctl	= compat_ptr_ioctl,
+#endif
+};
+
+struct evl_observable *evl_alloc_observable(int clone_flags)
+{
+	struct evl_observable *observable;
+	int ret;
+
+	observable = kzalloc(sizeof(*observable), GFP_KERNEL);
+	if (observable == NULL)
+		return ERR_PTR(-ENOMEM);
+
+	ret = evl_init_element(&observable->element,
+			&evl_observable_factory,
+			clone_flags | EVL_CLONE_OBSERVABLE);
+	if (ret) {
+		kfree(observable);
+		return ERR_PTR(ret);
+	}
+
+	INIT_LIST_HEAD(&observable->observers);
+	INIT_LIST_HEAD(&observable->flush_list);
+	evl_init_wait(&observable->oob_wait, &evl_mono_clock, EVL_WAIT_PRIO);
+	init_waitqueue_head(&observable->inband_wait);
+	init_irq_work(&observable->wake_irqwork, inband_wake_irqwork);
+	init_irq_work(&observable->flush_irqwork, inband_flush_irqwork);
+	evl_init_poll_head(&observable->poll_head);
+	evl_spin_lock_init(&observable->lock);
+	evl_index_factory_element(&observable->element);
+
+	return observable;
+}
+
+static struct evl_element *
+observable_factory_build(struct evl_factory *fac, const char *name,
+		void __user *u_attrs, int clone_flags, u32 *state_offp)
+{
+	struct evl_observable *observable;
+
+	if (clone_flags & ~EVL_OBSERVABLE_CLONE_FLAGS)
+		return ERR_PTR(-EINVAL);
+
+	observable = evl_alloc_observable(clone_flags);
+	if (IS_ERR(observable))
+		return ERR_PTR(PTR_ERR(observable));
+
+	return &observable->element;
+}
+
+static void observable_factory_dispose(struct evl_element *e)
+{
+	struct evl_observable *observable;
+
+	observable = container_of(e, struct evl_observable, element);
+	evl_destroy_wait(&observable->oob_wait);
+	evl_unindex_factory_element(&observable->element);
+	evl_destroy_element(&observable->element);
+	kfree_rcu(observable, element.rcu);
+}
+
+struct evl_factory evl_observable_factory = {
+	.name	=	EVL_OBSERVABLE_DEV,
+	.fops	=	&observable_fops,
+	.build =	observable_factory_build,
+	.dispose =	observable_factory_dispose,
+	/* Threads are observables in essence. */
+	.nrdev	=	CONFIG_EVL_NR_OBSERVABLES + CONFIG_EVL_NR_THREADS,
+	.flags	=	EVL_FACTORY_CLONE,
+};
diff --git a/kernel/evl/poll.c b/kernel/evl/poll.c
index 9502839552..ff4e8301e9 100644
--- a/kernel/evl/poll.c
+++ b/kernel/evl/poll.c
@@ -35,6 +35,7 @@ struct poll_group {
 struct poll_item {
 	unsigned int fd;
 	int events_polled;
+	union evl_value pollval;
 	struct rb_node rb;	    /* in group->item_index */
 	struct list_head next;	    /* in group->item_list */
 };
@@ -189,6 +190,7 @@ static int add_item(struct file *filp, struct poll_group *group,
 	item->fd = creq->fd;
 	events = creq->events & ~POLLNVAL;
 	item->events_polled = events | POLLERR | POLLHUP;
+	item->pollval = creq->pollval;
 
 	efilp = evl_get_file(creq->fd);
 	if (efilp == NULL) {
@@ -317,6 +319,7 @@ int mod_item(struct poll_group *group,
 	}
 
 	item->events_polled = events | POLLERR | POLLHUP;
+	item->pollval = creq->pollval;
 	new_generation(group);
 
 	evl_unlock_kmutex(&group->item_lock);
@@ -405,6 +408,7 @@ static int collect_events(struct poll_group *group,
 	list_for_each_entry(item, &group->item_list, next) {
 		wpt->fd = item->fd;
 		wpt->events_polled = item->events_polled;
+		wpt->pollval = item->pollval;
 		wpt++;
 	}
 
@@ -431,6 +435,7 @@ static int collect_events(struct poll_group *group,
 		ready &= wpt->events_polled | POLLNVAL;
 		if (ready) {
 			ev.fd = wpt->fd;
+			ev.pollval = wpt->pollval;
 			ev.events = ready;
 			ret = raw_copy_to_user(u_set, &ev, sizeof(ev));
 			if (ret)
diff --git a/kernel/evl/proxy.c b/kernel/evl/proxy.c
index 696ec5e5fd..05dd4849e2 100644
--- a/kernel/evl/proxy.c
+++ b/kernel/evl/proxy.c
@@ -315,6 +315,9 @@ proxy_factory_build(struct evl_factory *fac, const char *name,
 	size_t bufsz;
 	int ret;
 
+	if (clone_flags & ~EVL_CLONE_PUBLIC)
+		return ERR_PTR(-EINVAL);
+
 	ret = copy_from_user(&attrs, u_attrs, sizeof(attrs));
 	if (ret)
 		return ERR_PTR(-EFAULT);
@@ -374,7 +377,7 @@ proxy_factory_build(struct evl_factory *fac, const char *name,
 	evl_init_poll_head(&out->poll_head);
 	evl_init_flag(&out->oob_drained);
 	init_waitqueue_head(&out->inband_drained);
-	evl_index_element(&proxy->element);
+	evl_index_factory_element(&proxy->element);
 	mutex_init(&out->worker_lock);
 
 	return &proxy->element;
@@ -406,7 +409,7 @@ static void proxy_factory_dispose(struct evl_element *e)
 	}
 	fput(proxy->filp);
 	evl_destroy_flag(&out->oob_drained);
-	evl_unindex_element(&proxy->element);
+	evl_unindex_factory_element(&proxy->element);
 	evl_destroy_element(&proxy->element);
 
 	if (out->ring.bufmem)
diff --git a/kernel/evl/sched/core.c b/kernel/evl/sched/core.c
index a415a23bea..ff2fcc5995 100644
--- a/kernel/evl/sched/core.c
+++ b/kernel/evl/sched/core.c
@@ -96,7 +96,7 @@ static void watchdog_handler(struct evl_timer *timer) /* oob stage stalled */
 		curr->info |= T_KICKED;
 		evl_spin_unlock(&this_rq->lock);
 		evl_spin_unlock(&curr->lock);
-		evl_signal_thread(curr, SIGDEBUG, SIGDEBUG_WATCHDOG);
+		evl_notify_thread(curr, EVL_HMDIAG_WATCHDOG, evl_nil);
 		dovetail_send_mayday(current);
 		printk(EVL_WARNING "watchdog triggered on CPU #%d -- runaway thread "
 			"'%s' signaled\n", evl_rq_cpu(this_rq), curr->name);
@@ -189,6 +189,7 @@ static void init_rq(struct evl_rq *rq, int cpu)
 	 */
 	iattr.flags = T_ROOT;
 	iattr.affinity = cpumask_of(cpu);
+	iattr.observable = NULL;
 	iattr.sched_class = &evl_sched_idle;
 	iattr.sched_param.idle.prio = EVL_IDLE_PRIO;
 	evl_init_thread(&rq->root_thread, &iattr, rq, name_fmt, cpu);
@@ -1064,8 +1065,6 @@ EXPORT_SYMBOL_GPL(evl_switch_oob);
 void evl_switch_inband(int cause)
 {
 	struct evl_thread *curr = evl_current();
-	struct task_struct *p = current;
-	struct kernel_siginfo si;
 	struct evl_rq *this_rq;
 	bool notify;
 
@@ -1093,14 +1092,14 @@ void evl_switch_inband(int cause)
 
 	curr->state |= T_INBAND;
 	curr->local_info &= ~T_SYSRST;
-	notify = curr->state & T_USER && cause > SIGDEBUG_NONE;
+	notify = curr->state & T_USER && cause > EVL_HMDIAG_NONE;
 
 	/*
 	 * If we are initiating the ptsync sequence on breakpoint or
-	 * SIGSTOP/SIGINT is pending, do not send SIGDEBUG since
-	 * switching in-band is ok.
+	 * SIGSTOP/SIGINT is pending, do not send any HM notification
+	 * since switching in-band is ok.
 	 */
-	if (cause == SIGDEBUG_TRAP) {
+	if (cause == EVL_HMDIAG_TRAP) {
 		curr->info |= T_PTSTOP;
 		curr->info &= ~T_PTJOIN;
 		start_ptsync_locked(curr, this_rq);
@@ -1163,20 +1162,14 @@ void evl_switch_inband(int cause)
 	if (notify) {
 		/*
 		 * Help debugging spurious stage switches by sending
-		 * SIGDEBUG. We are running inband on the context of
-		 * the receiver, so we may bypass evl_signal_thread()
-		 * for this.
+		 * an HM event.
 		 */
-		if (curr->state & T_WOSS) {
-			memset(&si, 0, sizeof(si));
-			si.si_signo = SIGDEBUG;
-			si.si_code = SI_QUEUE;
-			si.si_int = cause | sigdebug_marker;
-			send_sig_info(SIGDEBUG, &si, p);
-		}
+		if (curr->state & T_WOSS)
+			evl_notify_thread(curr, cause, evl_nil);
+
 		/* May check for locking inconsistency too. */
 		if (curr->state & T_WOLI)
-			evl_detect_boost_drop(curr);
+			evl_detect_boost_drop();
 	}
 
 	/* @curr is now running inband. */
diff --git a/kernel/evl/stax.c b/kernel/evl/stax.c
index 91148d147a..c7927e8b6a 100644
--- a/kernel/evl/stax.c
+++ b/kernel/evl/stax.c
@@ -6,6 +6,7 @@
  * Stage exclusion lock implementation (aka STAX).
  */
 
+#include <evl/sched.h>
 #include <evl/stax.h>
 
 /* Gate marker for in-band activity. */
@@ -96,7 +97,7 @@ static int claim_stax_from_oob(struct evl_stax *stax, int gateval)
 	evl_spin_unlock_irqrestore(&stax->oob_wait.lock, flags);
 
 	if (notify) {
-		evl_signal_thread(curr, SIGDEBUG, SIGDEBUG_STAGE_LOCKED);
+		evl_notify_thread(curr, EVL_HMDIAG_STAGEX, evl_nil);
 		evl_kick_thread(curr, 0);
 	}
 
diff --git a/kernel/evl/syscall.c b/kernel/evl/syscall.c
index 4b90f21136..b66467ffc6 100644
--- a/kernel/evl/syscall.c
+++ b/kernel/evl/syscall.c
@@ -190,7 +190,7 @@ static void prepare_for_signal(struct task_struct *p,
 
 	evl_test_cancel();
 
-	evl_switch_inband(SIGDEBUG_MIGRATE_SIGNAL);
+	evl_switch_inband(EVL_HMDIAG_SIGDEMOTE);
 }
 
 /*
@@ -299,7 +299,7 @@ static int do_oob_syscall(struct irq_stage *stage, struct pt_regs *regs)
 			prepare_for_signal(p, curr, regs);
 		else if ((curr->state & T_WEAK) &&
 			!atomic_read(&curr->inband_disable_count))
-			evl_switch_inband(SIGDEBUG_NONE);
+			evl_switch_inband(EVL_HMDIAG_NONE);
 	}
 
 	/* Update the stats and user visible info. */
@@ -327,7 +327,7 @@ static int do_oob_syscall(struct irq_stage *stage, struct pt_regs *regs)
 	if (inband_syscall_nr(regs, &nr)) {
 		if (handle_vdso_fallback(nr, regs))
 			return SYSCALL_STOP;
-		evl_switch_inband(SIGDEBUG_MIGRATE_SYSCALL);
+		evl_switch_inband(EVL_HMDIAG_SYSDEMOTE);
 		return SYSCALL_PROPAGATE;
 	}
 
@@ -400,7 +400,7 @@ static int do_inband_syscall(struct irq_stage *stage, struct pt_regs *regs)
 			prepare_for_signal(p, curr, regs);
 		else if ((curr->state & T_WEAK) &&
 			!atomic_read(&curr->inband_disable_count))
-			evl_switch_inband(SIGDEBUG_NONE);
+			evl_switch_inband(EVL_HMDIAG_NONE);
 	}
 done:
 	if (curr->local_info & T_IGNOVR)
diff --git a/kernel/evl/thread.c b/kernel/evl/thread.c
index 8bb5075d4f..7b4d64d7c2 100644
--- a/kernel/evl/thread.c
+++ b/kernel/evl/thread.c
@@ -39,9 +39,14 @@
 #include <evl/mutex.h>
 #include <evl/poll.h>
 #include <evl/flag.h>
+#include <evl/factory.h>
+#include <evl/observable.h>
 #include <asm/evl/syscall.h>
 #include <trace/events/evl.h>
 
+#define EVL_THREAD_CLONE_FLAGS	\
+	(EVL_CLONE_PUBLIC|EVL_CLONE_OBSERVABLE|EVL_CLONE_MASTER)
+
 int evl_nrthreads;
 
 LIST_HEAD(evl_thread_list);
@@ -50,6 +55,8 @@ static DEFINE_HARD_SPINLOCK(thread_list_lock);
 
 static DECLARE_WAIT_QUEUE_HEAD(join_all);
 
+static const struct file_operations thread_fops;
+
 static void inband_task_wakeup(struct irq_work *work);
 
 static void skip_ptsync(struct evl_thread *thread);
@@ -171,6 +178,9 @@ int evl_init_thread(struct evl_thread *thread,
 	if ((flags & T_USER) && IS_ENABLED(CONFIG_EVL_DEBUG_WOLI))
 		flags |= T_WOLI;
 
+	if (iattr->observable)
+		flags |= T_OBSERV;
+
 	/*
 	 * If no rq was given, pick an initial CPU for the new thread
 	 * which is part of its affinity mask, and therefore also part
@@ -207,6 +217,7 @@ int evl_init_thread(struct evl_thread *thread,
 	thread->wwake = NULL;
 	thread->wait_data = NULL;
 	thread->u_window = NULL;
+	thread->observable = iattr->observable;
 	atomic_set(&thread->inband_disable_count, 0);
 	memset(&thread->poll_context, 0, sizeof(thread->poll_context));
 	memset(&thread->stat, 0, sizeof(thread->stat));
@@ -278,13 +289,17 @@ static void do_cleanup_current(struct evl_thread *curr)
 	unsigned long flags;
 	struct evl_rq *rq;
 
+	/* Kick out all subscribers. */
+	if (curr->observable)
+		evl_flush_observable(curr->observable);
+
 	/*
 	 * Drop trackers first since this may alter the rq state for
 	 * current.
 	 */
 	evl_drop_tracking_mutexes(curr);
 
-	evl_unindex_element(&curr->element);
+	evl_unindex_factory_element(&curr->element);
 
 	if (curr->state & T_USER) {
 		evl_free_chunk(&evl_shared_heap, curr->u_window);
@@ -448,7 +463,7 @@ int __evl_run_kthread(struct evl_kthread *kthread, int clone_flags)
 		goto fail_spawn;
 	}
 
-	evl_index_element(&thread->element);
+	evl_index_factory_element(&thread->element);
 	wait_for_completion(&kthread->done);
 	if (kthread->status)
 		return kthread->status;
@@ -983,7 +998,7 @@ int evl_join_thread(struct evl_thread *thread, bool uninterruptible)
 
 	if (curr && !(curr->state & T_INBAND)) {
 		evl_put_thread_rq(thread, rq, flags);
-		evl_switch_inband(SIGDEBUG_NONE);
+		evl_switch_inband(EVL_HMDIAG_NONE);
 		switched = true;
 	} else {
 		evl_put_thread_rq(thread, rq, flags);
@@ -1072,7 +1087,7 @@ void __evl_test_cancel(struct evl_thread *curr)
 		return;
 
 	if (!(curr->state & T_INBAND))
-		evl_switch_inband(SIGDEBUG_NONE);
+		evl_switch_inband(EVL_HMDIAG_NONE);
 
 	do_exit(0);
 	/* ... won't return ... */
@@ -1274,60 +1289,80 @@ void evl_demote_thread(struct evl_thread *thread)
 }
 EXPORT_SYMBOL_GPL(evl_demote_thread);
 
-struct inband_signal {
-	fundle_t fundle;
+struct sig_irqwork_data {
+	struct evl_thread *thread;
 	int signo, sigval;
 	struct irq_work work;
 };
 
-static void inband_task_signal(struct irq_work *work)
+static void do_inband_signal(struct evl_thread *thread, int signo, int sigval)
 {
-	struct evl_thread *thread;
-	struct inband_signal *req;
+	struct task_struct *p = thread->altsched.task;
 	struct kernel_siginfo si;
-	struct task_struct *p;
-	int signo;
 
-	req = container_of(work, struct inband_signal, work);
-	thread = evl_get_element_by_fundle(&evl_thread_factory,
-					req->fundle, struct evl_thread);
-	if (thread == NULL)
-		goto done;
-
-	p = thread->altsched.task;
-	signo = req->signo;
-	trace_evl_inband_signal(p, signo);
+	trace_evl_inband_signal(thread, signo, sigval);
 
 	if (signo == SIGDEBUG) {
 		memset(&si, '\0', sizeof(si));
 		si.si_signo = signo;
 		si.si_code = SI_QUEUE;
-		si.si_int = req->sigval;
+		si.si_int = sigval;
 		send_sig_info(signo, &si, p);
 	} else
 		send_sig(signo, p, 1);
+}
 
-	evl_put_element(&thread->element);
-done:
-	evl_free(req);
+static void sig_irqwork(struct irq_work *work)
+{
+	struct sig_irqwork_data *sigd;
+
+	sigd = container_of(work, struct sig_irqwork_data, work);
+	do_inband_signal(sigd->thread, sigd->signo, sigd->sigval);
+	evl_put_element(&sigd->thread->element);
+	evl_free(sigd);
 }
 
 void evl_signal_thread(struct evl_thread *thread, int sig, int arg)
 {
-	struct inband_signal *req;
+	struct sig_irqwork_data *sigd;
 
 	if (EVL_WARN_ON(CORE, !(thread->state & T_USER)))
 		return;
 
-	req = evl_alloc(sizeof(*req));
-	init_irq_work(&req->work, inband_task_signal);
-	req->fundle = fundle_of(thread);
-	req->signo = sig;
-	req->sigval = sig == SIGDEBUG ? arg | sigdebug_marker : arg;
-	irq_work_queue(&req->work);
+	if (running_inband()) {
+		do_inband_signal(thread, sig, arg);
+		return;
+	}
+
+	sigd = evl_alloc(sizeof(*sigd));
+	init_irq_work(&sigd->work, sig_irqwork);
+	sigd->thread = thread;
+	sigd->signo = sig;
+	sigd->sigval = sig == SIGDEBUG ? arg | sigdebug_marker : arg;
+
+	evl_get_element(&thread->element);
+	/* Cannot fail, irq_work is local to this call. */
+	irq_work_queue(&sigd->work);
 }
 EXPORT_SYMBOL_GPL(evl_signal_thread);
 
+void evl_notify_thread(struct evl_thread *thread,
+		int tag, union evl_value details)
+{
+	if (thread->state & T_HMSIG)
+		evl_signal_thread(thread, SIGDEBUG, tag);
+
+	if (thread->state & T_HMOBS) {
+		if (!evl_send_observable(thread->observable, tag, details))
+			printk_ratelimited(EVL_WARNING
+				"%s[%d] could not receive HM event #%d",
+				evl_element_name(&thread->element),
+					evl_get_inband_pid(thread),
+					tag);
+	}
+}
+EXPORT_SYMBOL_GPL(evl_notify_thread);
+
 #ifdef CONFIG_MMU
 
 static inline int commit_process_memory(void)
@@ -1503,7 +1538,7 @@ void handle_oob_trap_entry(unsigned int trapnr, struct pt_regs *regs)
 	 * We received a trap on the oob stage, switch to in-band
 	 * before handling the exception.
 	 */
-	evl_switch_inband(is_bp ? SIGDEBUG_TRAP : SIGDEBUG_MIGRATE_FAULT);
+	evl_switch_inband(is_bp ? EVL_HMDIAG_TRAP : EVL_HMDIAG_EXDEMOTE);
 }
 
 /* hard irqs on. */
@@ -1541,7 +1576,7 @@ void handle_oob_mayday(struct pt_regs *regs)
 	 * syscall. Filter this case out.
 	 */
 	if (!(curr->state & T_INBAND))
-		evl_switch_inband(SIGDEBUG_NONE);
+		evl_switch_inband(EVL_HMDIAG_NONE);
 }
 
 static void handle_migration_event(struct dovetail_migration_data *d)
@@ -1812,7 +1847,7 @@ static void handle_retuser_event(void)
 
 	if ((curr->state & T_WEAK) &&
 		atomic_read(&curr->inband_disable_count) == 0)
-		evl_switch_inband(SIGDEBUG_NONE);
+		evl_switch_inband(EVL_HMDIAG_NONE);
 }
 
 static void handle_cleanup_event(struct mm_struct *mm)
@@ -1848,7 +1883,9 @@ void handle_inband_event(enum inband_event_type event, void *data)
 		handle_sigwake_event(data);
 		break;
 	case INBAND_TASK_EXIT:
-		put_current_thread();
+		evl_drop_subscriptions(evl_get_subscriber());
+		if (evl_current())
+			put_current_thread();
 		break;
 	case INBAND_TASK_MIGRATION:
 		handle_migration_event(data);
@@ -2014,20 +2051,34 @@ static int update_mode(struct evl_thread *thread, __u32 mask,
 	unsigned long flags;
 	struct evl_rq *rq;
 
-	if (mask & ~(T_WOSS|T_WOLI|T_WOSX))
+	trace_evl_thread_update_mode(thread, mask, set);
+
+	if (mask & ~EVL_THREAD_MODE_BITS)
 		return -EINVAL;
 
-	trace_evl_thread_update_mode(thread, mask, set);
+	if (set) {
+		/* T_HMOBS requires observability of @thread. */
+		if (mask & T_HMOBS && thread->observable == NULL)
+			return -EINVAL;
+		/* Default to T_HMSIG if not specified. */
+		if (!(mask & (T_HMSIG|T_HMOBS)))
+			mask |= T_HMSIG;
+	}
 
 	rq = evl_get_thread_rq(thread, flags);
 
-	*oldmask = thread->state & (T_WOSS|T_WOLI|T_WOSX);
+	*oldmask = thread->state & EVL_THREAD_MODE_BITS;
 
 	if (likely(mask)) {
-		if (set)
+		if (set) {
 			thread->state |= mask;
-		else
+		} else {
 			thread->state &= ~mask;
+			if (!(thread->state & (T_WOSS|T_WOLI|T_WOSX)))
+				thread->state &= ~(T_HMSIG|T_HMOBS);
+			else if (!(thread->state & (T_HMSIG|T_HMOBS)))
+				thread->state &= ~(T_WOSS|T_WOLI|T_WOSX);
+		}
 	}
 
 	evl_put_thread_rq(thread, rq, flags);
@@ -2111,7 +2162,7 @@ static long thread_oob_ioctl(struct file *filp, unsigned int cmd,
 		break;
 	case EVL_THRIOC_SWITCH_INBAND:
 		if (thread == curr) {
-			evl_switch_inband(SIGDEBUG_NONE);
+			evl_switch_inband(EVL_HMDIAG_NONE);
 			ret = 0;
 		}
 		break;
@@ -2154,6 +2205,12 @@ static long thread_ioctl(struct file *filp, unsigned int cmd,
 	case EVL_THRIOC_JOIN:
 		ret = evl_join_thread(thread, false);
 		break;
+	case EVL_OBSIOC_SUBSCRIBE:
+	case EVL_OBSIOC_UNSUBSCRIBE:
+		if (thread->observable)
+			ret = evl_ioctl_observable(thread->observable,
+						cmd, arg);
+		break;
 	default:
 		ret = thread_common_ioctl(thread, cmd, arg);
 	}
@@ -2161,6 +2218,60 @@ static long thread_ioctl(struct file *filp, unsigned int cmd,
 	return ret;
 }
 
+static ssize_t thread_oob_read(struct file *filp,
+			char __user *u_buf, size_t count)
+{
+	struct evl_thread *thread = element_of(filp, struct evl_thread);
+
+	return evl_read_observable(thread->observable, u_buf, count,
+				!(filp->f_flags & O_NONBLOCK));
+}
+
+static ssize_t thread_oob_write(struct file *filp,
+			const char __user *u_buf, size_t count)
+{
+	struct evl_thread *thread = element_of(filp, struct evl_thread);
+
+	return evl_write_observable(thread->observable, u_buf, count);
+}
+
+static __poll_t thread_oob_poll(struct file *filp,
+				struct oob_poll_wait *wait)
+{
+	struct evl_thread *thread = element_of(filp, struct evl_thread);
+
+	return evl_oob_poll_observable(thread->observable, wait);
+}
+
+static ssize_t thread_write(struct file *filp, const char __user *u_buf,
+			size_t count, loff_t *ppos)
+{
+	struct evl_thread *thread = element_of(filp, struct evl_thread);
+
+	return evl_write_observable(thread->observable, u_buf, count);
+}
+
+static ssize_t thread_read(struct file *filp, char __user *u_buf,
+			size_t count, loff_t *ppos)
+{
+	struct evl_thread *thread = element_of(filp, struct evl_thread);
+
+	return evl_read_observable(thread->observable, u_buf, count,
+				!(filp->f_flags & O_NONBLOCK));
+}
+
+static __poll_t thread_poll(struct file *filp, poll_table *pt)
+{
+	struct evl_thread *thread = element_of(filp, struct evl_thread);
+
+	return evl_poll_observable(thread->observable, filp, pt);
+}
+
+bool evl_is_thread_file(struct file *filp)
+{
+	return filp->f_op == &thread_fops;
+}
+
 static const struct file_operations thread_fops = {
 	.open		= evl_open_element,
 	.release	= evl_release_element,
@@ -2170,6 +2281,13 @@ static const struct file_operations thread_fops = {
 	.compat_ioctl   = compat_ptr_ioctl,
 	.compat_oob_ioctl  = compat_ptr_oob_ioctl,
 #endif
+	/* For observability. */
+	.oob_read	= thread_oob_read,
+	.oob_write	= thread_oob_write,
+	.oob_poll	= thread_oob_poll,
+	.read		= thread_read,
+	.write		= thread_write,
+	.poll		= thread_poll,
 };
 
 static int map_uthread_self(struct evl_thread *thread)
@@ -2236,6 +2354,7 @@ static int map_uthread_self(struct evl_thread *thread)
 /*
  * Deconstruct a thread we just failed to map over a userland task.
  * Since the former must be dormant, it can't be part of any runqueue.
+ * The caller is in charge of freeing @thread.
  */
 static void discard_unmapped_uthread(struct evl_thread *thread)
 {
@@ -2245,14 +2364,13 @@ static void discard_unmapped_uthread(struct evl_thread *thread)
 
 	if (thread->u_window)
 		evl_free_chunk(&evl_shared_heap, thread->u_window);
-
-	kfree(thread);
 }
 
 static struct evl_element *
 thread_factory_build(struct evl_factory *fac, const char *name,
 		void __user *u_attrs, int clone_flags, u32 *state_offp)
 {
+	struct evl_observable *observable = NULL;
 	struct task_struct *tsk = current;
 	struct evl_init_thread_attr iattr;
 	struct evl_thread *curr;
@@ -2261,6 +2379,9 @@ thread_factory_build(struct evl_factory *fac, const char *name,
 	if (evl_current())
 		return ERR_PTR(-EBUSY);
 
+	if (clone_flags & ~EVL_THREAD_CLONE_FLAGS)
+		return ERR_PTR(-EINVAL);
+
 	/* @current must open the control device first. */
 	if (!test_bit(EVL_MM_ACTIVE_BIT, &dovetail_mm_state()->flags))
 		return ERR_PTR(-EPERM);
@@ -2271,31 +2392,49 @@ thread_factory_build(struct evl_factory *fac, const char *name,
 
 	ret = evl_init_element(&curr->element,
 			&evl_thread_factory, clone_flags);
-	if (ret) {
-		kfree(curr);
-		return ERR_PTR(ret);
-	}
+	if (ret)
+		goto fail_element;
 
 	iattr.flags = T_USER;
+
+	if (clone_flags & EVL_CLONE_OBSERVABLE) {
+		/*
+		 * Accessing the observable is done via the thread
+		 * element (if public), so clear the public flag for
+		 * the observable itself.
+		 */
+		observable = evl_alloc_observable(
+			clone_flags & ~EVL_CLONE_PUBLIC);
+		if (IS_ERR(observable)) {
+			ret = PTR_ERR(observable);
+			goto fail_observable;
+		}
+		ret = evl_create_core_element_device(
+			&observable->element,
+			&evl_observable_factory,
+			name);
+		if (ret)
+			goto fail_observable_dev;
+		observable = observable;
+	} else if (clone_flags & EVL_CLONE_MASTER) {
+		ret = -EINVAL;
+		goto fail_observable;
+	}
+
 	iattr.affinity = cpu_possible_mask;
+	iattr.observable = observable;
 	iattr.sched_class = &evl_sched_weak;
 	iattr.sched_param.weak.prio = 0;
 	ret = evl_init_thread(curr, &iattr, NULL, "%s", name);
-	if (ret) {
-		evl_destroy_element(&curr->element);
-		kfree(curr);
-		return ERR_PTR(ret);
-	}
+	if (ret)
+		goto fail_thread;
 
 	ret = map_uthread_self(curr);
-	if (ret) {
-		evl_destroy_element(&curr->element);
-		discard_unmapped_uthread(curr);
-		return ERR_PTR(ret);
-	}
+	if (ret)
+		goto fail_map;
 
 	*state_offp = evl_shared_offset(curr->u_window);
-	evl_index_element(&curr->element);
+	evl_index_factory_element(&curr->element);
 
 	/*
 	 * Unlike most elements, a thread may exist in absence of any
@@ -2311,6 +2450,19 @@ thread_factory_build(struct evl_factory *fac, const char *name,
 	tsk->comm[sizeof(tsk->comm) - 1] = '\0';
 
 	return &curr->element;
+
+fail_map:
+	discard_unmapped_uthread(curr);
+fail_thread:
+	if (observable)
+fail_observable_dev:
+		evl_put_element(&observable->element); /* ->dispose() */
+fail_observable:
+	evl_destroy_element(&curr->element);
+fail_element:
+	kfree(curr);
+
+	return ERR_PTR(ret);
 }
 
 static void thread_factory_dispose(struct evl_element *e)
@@ -2330,6 +2482,8 @@ static void thread_factory_dispose(struct evl_element *e)
 	 * and no more reachable, so we can wakeup joiners if any.
 	 */
 	if (likely(state & T_ZOMBIE)) {
+		if (thread->observable)
+			evl_put_element(&thread->observable->element);
 		evl_destroy_element(&thread->element);
 		complete_all(&thread->exited);	 /* evl_join_thread() */
 		if (waitqueue_active(&join_all)) /* evl_killall() */
@@ -2337,6 +2491,8 @@ static void thread_factory_dispose(struct evl_element *e)
 	} else {
 		if (EVL_WARN_ON(CORE, evl_current() != thread))
 			return;
+		if (thread->observable)
+			evl_put_element(&thread->observable->element);
 		cleanup_current_thread();
 		evl_destroy_element(&thread->element);
 	}
@@ -2521,12 +2677,33 @@ static ssize_t pid_show(struct device *dev,
 }
 static DEVICE_ATTR_RO(pid);
 
+static ssize_t observable_show(struct device *dev,
+			struct device_attribute *attr,
+			char *buf)
+{
+	struct evl_thread *thread;
+	ssize_t ret;
+
+	thread = evl_get_element_by_dev(dev, struct evl_thread);
+	if (thread == NULL)
+		return -EIO;
+
+	ret = snprintf(buf, PAGE_SIZE, "%d\n",
+		evl_element_is_observable(&thread->element));
+
+	evl_put_element(&thread->element);
+
+	return ret;
+}
+static DEVICE_ATTR_RO(observable);
+
 static struct attribute *thread_attrs[] = {
 	&dev_attr_state.attr,
 	&dev_attr_sched.attr,
 	&dev_attr_timeout.attr,
 	&dev_attr_stats.attr,
 	&dev_attr_pid.attr,
+	&dev_attr_observable.attr,
 	NULL,
 };
 ATTRIBUTE_GROUPS(thread);
diff --git a/kernel/evl/wait.c b/kernel/evl/wait.c
index d3a3e0a7fd..0314774598 100644
--- a/kernel/evl/wait.c
+++ b/kernel/evl/wait.c
@@ -45,7 +45,7 @@ void evl_add_wait_queue(struct evl_wait_queue *wq, ktime_t timeout,
 
 	if ((curr->state & T_WOLI) &&
 		atomic_read(&curr->inband_disable_count) > 0)
-		evl_signal_thread(curr, SIGDEBUG, SIGDEBUG_MUTEX_SLEEP);
+		evl_notify_thread(curr, EVL_HMDIAG_LKSLEEP, evl_nil);
 
 	if (!(wq->flags & EVL_WAIT_PRIO))
 		list_add_tail(&curr->wait_next, &wq->wchan.wait_list);
@@ -205,3 +205,11 @@ int evl_wait_schedule(struct evl_wait_queue *wq)
 	return ret;
 }
 EXPORT_SYMBOL_GPL(evl_wait_schedule);
+
+struct evl_thread *evl_wait_head(struct evl_wait_queue *wq)
+{
+	assert_evl_lock(&wq->lock);
+	return list_first_entry_or_null(&wq->wchan.wait_list,
+					struct evl_thread, wait_next);
+}
+EXPORT_SYMBOL_GPL(evl_wait_head);
diff --git a/kernel/evl/xbuf.c b/kernel/evl/xbuf.c
index aaa37032da..8e8d427538 100644
--- a/kernel/evl/xbuf.c
+++ b/kernel/evl/xbuf.c
@@ -639,6 +639,9 @@ xbuf_factory_build(struct evl_factory *fac, const char *name,
 	if (ret)
 		return ERR_PTR(-EFAULT);
 
+	if (clone_flags & ~EVL_CLONE_PUBLIC)
+		return ERR_PTR(-EINVAL);
+
 	/* LART */
 	if ((attrs.i_bufsz == 0 && attrs.o_bufsz == 0) ||
 		order_base_2(attrs.i_bufsz) > 30 ||
-- 
2.30.2

